<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Visualizing copy number data as heatmaps • GAMBLR.viz</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="CopyNumberHeatmap_files/libs/quarto-html/popper.min.js"></script><script src="CopyNumberHeatmap_files/libs/quarto-html/tippy.umd.min.js"></script><link href="CopyNumberHeatmap_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="CopyNumberHeatmap_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Visualizing copy number data as heatmaps">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-viz.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Installation</h6></li>
    <li><a class="dropdown-item" href="../articles/install.html">Installation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/recap_figures.html">Reproducing figures from GAMBL papers</a></li>
    <li><a class="dropdown-item" href="../articles/forestplot.html">Easy forest plots</a></li>
    <li><a class="dropdown-item" href="../articles/CopyNumberHeatmap.html">Visualizing copy number data as heatmaps</a></li>
    <li><a class="dropdown-item" href="../articles/data_q.html">Exploring data quality</a></li>
    <li><a class="dropdown-item" href="../articles/oncoplot.html">Upgrade your Oncoplots</a></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Metadata and GAMBLR colours</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Help</h6></li>
    <li><a class="dropdown-item" href="../articles/faq.html">Frequently Asked Qestions</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Visualizing copy number data as heatmaps</h1>





      <div class="d-none name"><code></code></div>
    </div>






    <p>There is a wealth of copy number data bundled within <code>GAMBLR.data</code> in the form of segmented copy number profiles. This tutorial will give you an overview of this data. It will also demonstrate how to process segmented data into a set of bins such that the data can be represented as a matrix of CN values or log ratios.</p>
<section class="level2"><h2 id="prepare-setup">Prepare setup<a class="anchor" aria-label="anchor" href="#prepare-setup"></a>
</h2>
<p>We will first import the necessary packages:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level2"><h2 id="metadata">Metadata<a class="anchor" aria-label="anchor" href="#metadata"></a>
</h2>
<p>As usual, we typically begin an analysis by loading the metadata and restricting the rows to samples of interest based on any number of criteria. Here, we’ll separately look at the data from WGS (<code>genome</code> seq_type) and whole exome sequencing <code>capture</code> seq_type.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dlbcl_genome_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">==</span><span class="st">"DLBCL"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dlbcl_exome_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span>seq_type_filter <span class="op">=</span> <span class="st">"capture"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">==</span><span class="st">"DLBCL"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#remove any duplicate sample_id/seq_type combinations</span></span>
<span><span class="va">g_meta_clean</span> <span class="op">=</span> <span class="fu">check_and_clean_metadata</span><span class="op">(</span><span class="va">dlbcl_genome_meta</span>,</span>
<span>                                      duplicate_action <span class="op">=</span> <span class="st">"keep_first"</span><span class="op">)</span></span>
<span><span class="co">#remove any duplicate sample_id/seq_type combinations</span></span>
<span><span class="va">e_meta_clean</span> <span class="op">=</span> <span class="fu">check_and_clean_metadata</span><span class="op">(</span><span class="va">dlbcl_exome_meta</span>,</span>
<span>                                      duplicate_action <span class="op">=</span> <span class="st">"keep_first"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">g_meta_clean</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 529</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">e_meta_clean</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1783</code></pre>
</div>
</div>
<p>This shows that we have data from 529 genomes and 1783 exomes just from DLBCLs.</p>
</section><section class="level2"><h2 id="projections">Projections<a class="anchor" aria-label="anchor" href="#projections"></a>
</h2>
<p>The number of samples that have copy number results available isn’t necessarily going to match this. Importantly, some studies are older and relied on the <code>grch37</code> (i.e. hg19) genome build whereas other studies used some flavour of <code>hg38</code>. In GAMBL, we aim to <em>project</em> all results to both <code>grch37</code> and <code>hg38</code> so results are relatively comparable between studies. However, for efficiency, GAMBLR.data doesn’t have both projections for every data set (sorry!). Let’s start by determining which of our samples in the metadata have data available for in case there are some results only available for one or the other <em>projection</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">g_segments_grch37</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_cn_segments.html" class="external-link">get_cn_segments</a></span><span class="op">(</span>these <span class="op">=</span> <span class="va">g_meta_clean</span>,</span>
<span>                                    projection <span class="op">=</span> <span class="st">'grch37'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">g_segments_hg38</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_cn_segments.html" class="external-link">get_cn_segments</a></span><span class="op">(</span>these <span class="op">=</span> <span class="va">g_meta_clean</span>,</span>
<span>                                    projection <span class="op">=</span> <span class="st">'hg38'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">g_segments_grch37</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 214</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">g_segments_hg38</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 48</code></pre>
</div>
</div>
<p>We seem to only have copy number segments available from 48 samples for hg38. For grch37 we have data from 214 samples. How about the exome data?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">e_segments_grch37</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_cn_segments.html" class="external-link">get_cn_segments</a></span><span class="op">(</span>these <span class="op">=</span> <span class="va">e_meta_clean</span>,</span>
<span>                                    projection <span class="op">=</span> <span class="st">'grch37'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">e_segments_hg38</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_cn_segments.html" class="external-link">get_cn_segments</a></span><span class="op">(</span>these <span class="op">=</span> <span class="va">e_meta_clean</span>,</span>
<span>                                    projection <span class="op">=</span> <span class="st">'hg38'</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">e_segments_grch37</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">e_segments_hg38</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
</div>
<p>Ouch! We only provided copy number data from <code>genome</code> samples (so far). Let’s proceed with the largest data set, which is the <code>genome</code> samples and grch37 projection.</p>
</section><section class="level2"><h2 id="from-segments-to-bins">From segments to bins<a class="anchor" aria-label="anchor" href="#from-segments-to-bins"></a>
</h2>
<p>Segmented copy number data can be difficult to work with for a variety of reasons. For any coordinate in the genome, it’s unnecessarily complicated to determine the copy number of all samples. We simplify this by breaking the genome into a series of <em>bins</em> and determining the copy number state of each bin across all samples. This yields a N by M matrix where N is the number of samples with copy number data and M is the number of bins. The <code>segmented_data_to_cn_matrix</code> function in <code>GAMBLR.utils</code> makes this easy. In this example, we’ll do this to our 214 using 2500 bins. I’ve provided the metadata here even though it’s not strictly necessary. This can be useful if you pass seg_data that contains rows from patients that you don’t plan to analyze. Those rows will be removed before the slow process begins.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">time_start</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span>;</span>
<span><span class="va">dlbcl_cn_binned</span> <span class="op">=</span> <span class="fu">segmented_data_to_cn_matrix</span><span class="op">(</span></span>
<span>                                  seg_data <span class="op">=</span> <span class="va">g_segments_grch37</span>,</span>
<span>                                  strategy<span class="op">=</span><span class="st">"auto_split"</span>,</span>
<span>                                  n_bins_split<span class="op">=</span><span class="fl">2500</span>,</span>
<span>                                  these_samples_metadata <span class="op">=</span> <span class="va">g_meta_clean</span><span class="op">)</span></span>
<span><span class="va">time_end</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span>;</span>
<span><span class="va">elapsed</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">time_end</span> <span class="op">-</span> <span class="va">time_start</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">elapsed</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Time difference of 18 secs</code></pre>
</div>
</div>
<p>That took 18 to finish. As you might imagine, increasing the n_bins_split value will make this process slower but will increase the resolution of your analysis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">every_seg</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_cn_segments.html" class="external-link">get_cn_segments</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">every_seg</span><span class="op">$</span><span class="va">ID</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 448</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">time_start</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span>;</span>
<span><span class="va">all_cn_binned</span> <span class="op">=</span> <span class="fu">segmented_data_to_cn_matrix</span><span class="op">(</span></span>
<span>                                  seg_data <span class="op">=</span> <span class="va">every_seg</span>,</span>
<span>                                  strategy<span class="op">=</span><span class="st">"auto_split"</span>,</span>
<span>                                  n_bins_split<span class="op">=</span><span class="fl">2500</span><span class="op">)</span> <span class="co">#no metadata provided</span></span>
<span><span class="va">time_end</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span>;</span>
<span><span class="va">elapsed</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">time_end</span> <span class="op">-</span> <span class="va">time_start</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">elapsed</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Time difference of 25 secs</code></pre>
</div>
</div>
<p>As you can see, the process took considerably longer this time because of the additional samples that weren’t dropped.</p>
<p>As you might have guessed from the <code>strategy</code> parameter, there are other strategies for converting segmented data into a copy number matrix that don’t rely on equal-sized bins. The other options are ‘custom_regions’,‘cytobands’ and ‘GISTIC’. Each of these will be covered in another tutorial. You can find out more about these in the <code>segmented_data_to_cn_matrix</code> documentation.</p>
</section><section class="level2"><h2 id="dlbcl-copy-number-heatmap">DLBCL Copy Number Heatmap<a class="anchor" aria-label="anchor" href="#dlbcl-copy-number-heatmap"></a>
</h2>
<p>Let’s start with the most basic example. Using the copy number matrix we just created, we run <code>pretty_CN_heatmap</code>. As for most other functions, you have to provide your metadata table via <em>these_samples_metadata</em>. You can often unambiguously refer to this parameter with the word <em>these</em>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate a basic genome-wide CN heatmap</span></span>
<span><span class="fu"><a href="../reference/pretty_CN_heatmap.html">pretty_CN_heatmap</a></span><span class="op">(</span>cn_state_matrix<span class="op">=</span><span class="va">dlbcl_cn_binned</span>,</span>
<span>                  these_samples_metadata <span class="op">=</span> <span class="va">g_meta_clean</span>,</span>
<span>                  hide_annotations <span class="op">=</span> <span class="st">"chromosome"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="CopyNumberHeatmap_files/figure-html/dlbcl_heatmap-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 id="copy-number-heatmap-for-everything">Copy Number Heatmap for everything<a class="anchor" aria-label="anchor" href="#copy-number-heatmap-for-everything"></a>
</h2>
<p>In an earlier example, we generated a larger matrix from all the <code>grch37</code> copy number data in <code>GAMBLR.data</code>. There are <code>nrow(all_cn_binned)</code> samples in there. Let’s see what that looks like in a heatmap. This time, we’ll tell the function to scale the copy number values to the average ploidy of each sample using <code>scale_by_sample</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">all_genome_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">all_genome_meta</span> <span class="op">=</span> <span class="fu">check_and_clean_metadata</span><span class="op">(</span><span class="va">all_genome_meta</span>,</span>
<span>                                      duplicate_action <span class="op">=</span> <span class="st">"keep_first"</span><span class="op">)</span></span>
<span><span class="co">#drop any genomes that don't have copy number data</span></span>
<span></span>
<span><span class="va">all_genome_meta</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">all_genome_meta</span>,</span>
<span>  <span class="va">sample_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">all_cn_binned</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co"># Generate a genome-wide CN heatmap without clustering</span></span>
<span><span class="co"># Instead, we'll order samples based on pathology and lymphgen</span></span>
<span><span class="fu"><a href="../reference/pretty_CN_heatmap.html">pretty_CN_heatmap</a></span><span class="op">(</span>cn_state_matrix<span class="op">=</span><span class="va">all_cn_binned</span>,</span>
<span>                  these_samples_metadata <span class="op">=</span> <span class="va">all_genome_meta</span>,</span>
<span>                  scale_by_sample <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                  cluster_rows<span class="op">=</span><span class="cn">F</span>,</span>
<span>                  metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>,<span class="st">"lymphgen"</span><span class="op">)</span>,</span>
<span>                  sortByMetadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>,<span class="st">"lymphgen"</span><span class="op">)</span>,</span>
<span>                  bin_label_nudge <span class="op">=</span> <span class="fl">1.08</span>,</span>
<span>                  labelTheseGenes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDKN2A"</span>,<span class="st">"TP53"</span><span class="op">)</span>,</span>
<span>                  hide_annotations <span class="op">=</span> <span class="st">"chromosome"</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="CopyNumberHeatmap_files/figure-html/all_heatmap-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'light-border',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config);
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
