<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Exploring data quality • GAMBLR.viz</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="data_q_files/libs/quarto-html/popper.min.js"></script><script src="data_q_files/libs/quarto-html/tippy.umd.min.js"></script><link href="data_q_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="data_q_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Exploring data quality">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Introduction</h6></li>
    <li><a class="dropdown-item" href="../articles/why.html">GAMBLR.viz</a></li>
    <li><a class="dropdown-item" href="../articles/install.html"></a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/recap_figures.html">Reproducing figures from GAMBL papers</a></li>
    <li><a class="dropdown-item" href="../articles/forestplot.html">Easy forest plots</a></li>
    <li><a class="dropdown-item" href="../articles/CopyNumberHeatmap.html">Visualizing copy number data as heatmaps</a></li>
    <li><a class="dropdown-item" href="../articles/data_q.html">Exploring data quality</a></li>
    <li><a class="dropdown-item" href="../articles/oncoplot.html">Upgrade your Oncoplots</a></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Metadata and GAMBLR colours</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Help</h6></li>
    <li><a class="dropdown-item" href="../articles/faq.html">Frequently Asked Qestions</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Exploring data quality</h1>





      <div class="d-none name"><code></code></div>
    </div>






    <p>When comparing the mutation data from different sources, it is important to consider different underlying biases that may lead to unexpected results or incorrect conclusions. Different centers and groups may have different acceptable standards to library preparation, sequencing, data quality control, variant calling and analysis. More discussion on this can be found in the <a href="https://doi.org/10.1182/blood.2022016095" class="external-link">Dreval et al</a> Blood spotlight published in 2022. <br> Can we use GAMBLR.viz to conduct similar investigation and take advantage of the bundled data and the visualizations available with this package? <br> This tutorial will demonstate the example of the inputs and showcase the main features of such analysis, replicating some panels and figures from that study.</p>
<section class="level2"><h2 id="prepare-setup">Prepare setup<a class="anchor" aria-label="anchor" href="#prepare-setup"></a>
</h2>
<p>We will first import the necessary packages:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eclarke/ggbeeswarm" class="external-link">ggbeeswarm</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/daattali/ggExtra" class="external-link">ggExtra</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Next, we will get some data to display. We have explored before how to obrain data bundled with <a href="https://morinlab.github.io/GAMBLR.data/tutorials/getting_started.html" class="external-link">GAMBLR.data</a> and the difference between the <a href="https://morinlab.github.io/GAMBLR.data/resources/bundled_data.html" class="external-link">available objects</a>, so here we will use the demonstrated approaches to retreive metadata and simple somatic mutations. For demonstration purposes, we will focus on the largest <a href="https://morinlab.github.io/GAMBLR.data/faq.html#what-are-the-studies-bundled-with-the-package" class="external-link">whole exome sequencing studies</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get metadata</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">gambl_metadata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">cohort</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dlbcl_chapuy"</span>, <span class="st">"dlbcl_reddy"</span>, <span class="st">"dlbcl_schmitz"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># Get mutations</span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_coding_ssm.html" class="external-link">get_coding_ssm</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    this_seq_type <span class="op">=</span> <span class="st">"capture"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<p>Next, we can take advantage of the <code><a href="https://rdrr.io/pkg/GAMBLR.open/man/collate_results.html" class="external-link">collate_results()</a></code> function in GAMBLR.data to add all the necessary QC metrics that were directly reported in the <a href="https://doi.org/10.1182/blood.2022016095" class="external-link">MIRAGE manuscript.</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/collate_results.html" class="external-link">collate_results</a></span><span class="op">(</span></span>
<span>    sample_table <span class="op">=</span> <span class="va">metadata</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">metadata</span>, <span class="va">.</span>, by <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Now we have our metadata and mutations we want to explore, so we are ready to start exploring and visualizing the data.</p>
</section><section class="level2"><h2 id="target-space-coverage-across-the-studies">Target space coverage across the studies<a class="anchor" aria-label="anchor" href="#target-space-coverage-across-the-studies"></a>
</h2>
<p>The original manuscript highlights a striking difference in the effective nonredundant coverage across the target space (MeanCorrectedCoverage). Let’s see how that plot was generated:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="va">study</span>,</span>
<span>            y <span class="op">=</span> <span class="va">MeanCorrectedCoverage</span>,</span>
<span>            colour <span class="op">=</span> <span class="va">ffpe_or_frozen</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>outlier.shape <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_quasirandom.html" class="external-link">geom_quasirandom</a></span><span class="op">(</span></span>
<span>        dodge.width <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>        alpha <span class="op">=</span> <span class="fl">0.5</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span></span>
<span>        <span class="op">~</span><span class="va">seq_type</span>,</span>
<span>        ncol <span class="op">=</span> <span class="fl">1</span>,</span>
<span>        scales <span class="op">=</span> <span class="st">"free"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_colour_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu">get_gambl_colours</span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_Morons</span><span class="op">(</span></span>
<span>        my_legend_position <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>        my_legend_direction <span class="op">=</span> <span class="st">"vertical"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="data_q_files/figure-html/cov-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Indeed, the coverage in one of the studies is very different from the rest. This plot, also shown on panel B of Figure 1 in the manuscript, also shows that there are only 5/999 samples in the study by Reddy et al that have coverage figher the sample with the lowest coverage in the study from Schmitz et al! Seeing this, the logical question arises: does this difference in the sample quality affects the ability to detect coding mutations in lymphoma genes?</p>
</section><section class="level2"><h2 id="correlation-between-coverage-and-mutation-discovery">Correlation between coverage and mutation discovery<a class="anchor" aria-label="anchor" href="#correlation-between-coverage-and-mutation-discovery"></a>
</h2>
<p>Next, we can see whether there is any limitation to detect somatic mutations in the relevant genes when the sample is of a very poor quality. Indeed, the plot below replicates the panel C of Figure 1 in the manuscript and shows that there is direct effect:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># reformat naming of the standard colors</span></span>
<span><span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu">get_gambl_colours</span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Schmitz"</span>, <span class="st">"Chapuy"</span>, <span class="st">"Reddy"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Schmitz, 2018"</span>, <span class="st">"Chapuy, 2018"</span>, <span class="st">"Reddy, 2017"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="va">MeanCorrectedCoverage</span>,</span>
<span>            y <span class="op">=</span> <span class="va">coding_mutations</span>,</span>
<span>            colour <span class="op">=</span> <span class="va">study</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">200</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_Morons</span><span class="op">(</span></span>
<span>        base_size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>        my_legend_position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>        my_legend_direction <span class="op">=</span> <span class="st">"horizontal"</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># display density</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ggExtra/man/ggMarginal.html" class="external-link">ggMarginal</a></span><span class="op">(</span></span>
<span>    <span class="va">p</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"density"</span>,</span>
<span>    groupColour <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    groupFill <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="data_q_files/figure-html/unnamed-chunk-1-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 id="find-significant-differences-between-low-and-high-coverage-groups">Find significant differences between low and high coverage groups<a class="anchor" aria-label="anchor" href="#find-significant-differences-between-low-and-high-coverage-groups"></a>
</h2>
<p>Next, we can explore whether or not there are any significant differences in the frequency of mutations as detected in those of high and low coverage. As a cutoff, we arbitrarily selected a value of 25X for the sample to be considered in of the “High” or “Low” groups. We will here reformat the metadata to a more convenient format:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mirage_meta</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        good_coverage <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">MeanCorrectedCoverage</span> <span class="op">&gt;</span> <span class="fl">25</span> , <span class="st">"TRUE"</span>, <span class="st">"FALSE"</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="va">as.data.frame</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        sample_id <span class="op">=</span> <span class="va">Tumor_Sample_Barcode</span>,</span>
<span>        `DNA source` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span></span>
<span>            <span class="va">ffpe_or_frozen</span> <span class="op">==</span> <span class="st">"ffpe"</span>,</span>
<span>            <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">toupper</a></span><span class="op">(</span><span class="va">ffpe_or_frozen</span><span class="op">)</span>,</span>
<span>            <span class="va">ffpe_or_frozen</span></span>
<span>        <span class="op">)</span>,</span>
<span>        Pathology <span class="op">=</span> <span class="va">pathology</span>,</span>
<span>        Study <span class="op">=</span> <span class="va">study</span>,</span>
<span>        `Pairing status` <span class="op">=</span> <span class="va">pairing_status</span>,</span>
<span>        Protocol <span class="op">=</span> <span class="va">seq_type</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<p>For the demonstation purposes, let’s focus only on a set of genes reported as significantly mutated in the Reddy et al study. To find out these genes, we do not need search Pubmed for the gene lists, but can convently retreive from one of the <a href="https://morinlab.github.io/GAMBLR.data/resources/bundled_data.html" class="external-link">GAMBLR.data objects</a>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GAMBLR.data/man/reddy_genes.html" class="external-link">reddy_genes</a></span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">hgnc_symbol</span><span class="op">)</span></span></code></pre></div>
</div>
<p>We now can take a look at what genes are significantly different between high and low coverage groups:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">forest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyForestPlot.html">prettyForestPlot</a></span><span class="op">(</span></span>
<span>    maf <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    metadata <span class="op">=</span> <span class="va">mirage_meta</span>,</span>
<span>    comparison_column <span class="op">=</span> <span class="st">"good_coverage"</span>,</span>
<span>    max_q <span class="op">=</span> <span class="fl">0.1</span></span>
<span><span class="op">)</span></span>
<span><span class="va">forest</span><span class="op">$</span><span class="va">arranged</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="data_q_files/figure-html/fdr_plot-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
<p>That’s a lot of difference! We can easily display this information in the form of oncoplot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># display genes in the order of decreasing difference</span></span>
<span><span class="va">gene_order</span> <span class="op">&lt;-</span> <span class="va">forest</span><span class="op">$</span><span class="va">fisher</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">estimate</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="../reference/prettyCoOncoplot.html">prettyCoOncoplot</a></span><span class="op">(</span></span>
<span>    metadata <span class="op">=</span> <span class="va">mirage_meta</span>,</span>
<span>    maf <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    comparison_column <span class="op">=</span> <span class="st">"good_coverage"</span>,</span>
<span>    label1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>        <span class="st">"High coverage (&gt;25X, N="</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mirage_meta</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">good_coverage</span> <span class="op">==</span> <span class="st">"TRUE"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="st">")"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    label2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span></span>
<span>        <span class="st">"Low coverage (&lt;25X, N="</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">mirage_meta</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">good_coverage</span> <span class="op">==</span> <span class="st">"FALSE"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="st">")"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>        <span class="st">"Pairing status"</span>,</span>
<span>        <span class="st">"DNA source"</span>,</span>
<span>        <span class="st">"Protocol"</span>,</span>
<span>        <span class="st">"Study"</span>,</span>
<span>        <span class="st">"MeanCorrectedCoverage"</span></span>
<span>    <span class="op">)</span>,</span>
<span>    numericMetadataColumns <span class="op">=</span> <span class="st">"MeanCorrectedCoverage"</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="st">"MeanCorrectedCoverage"</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="fl">12</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="fl">7</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">gene_order</span>,</span>
<span>    keepGeneOrder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    arrange_descending <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    legend_row <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    annotation_row <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="data_q_files/figure-html/coonco-1.png" width="1440"></p>
</figure>
</div>
</div>
</div>
</section><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'light-border',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config);
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
