<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Easy forest plots â€¢ GAMBLR.viz</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="forestplot_files/libs/quarto-html/popper.min.js"></script><script src="forestplot_files/libs/quarto-html/tippy.umd.min.js"></script><link href="forestplot_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="forestplot_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Easy forest plots">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Introduction</h6></li>
    <li><a class="dropdown-item" href="../articles/why.html">GAMBLR.viz</a></li>
    <li><a class="dropdown-item" href="../articles/install.html"></a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/recap_figures.html">Reproducing figures from GAMBL papers</a></li>
    <li><a class="dropdown-item" href="../articles/forestplot.html">Easy forest plots</a></li>
    <li><a class="dropdown-item" href="../articles/CopyNumberHeatmap.html">Visualizing copy number data as heatmaps</a></li>
    <li><a class="dropdown-item" href="../articles/data_q.html">Exploring data quality</a></li>
    <li><a class="dropdown-item" href="../articles/oncoplot.html">Upgrade your Oncoplots</a></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Metadata and GAMBLR colours</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Help</h6></li>
    <li><a class="dropdown-item" href="../articles/faq.html">Frequently Asked Qestions</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Easy forest plots</h1>





      <div class="d-none name"><code></code></div>
    </div>






    <p>This tutorial introduces one of the ways GAMBLR.viz facilitates exploration of mutation incidence differences between groups of samples. Offering our take on forest plots, <code>prettyForestPlot</code> is part of a group of similarly named GAMBLR.viz functions. There is no specific formatting or data preparation needed for the analysis and visualization, and the only required inputs are the mutation data (can be maf format or binary feature matrix), metadata (containing sample identifiers in <code>sample_id</code> column and annotation of the group that will be used in comparison), and a character of the column name in metadata where the sample annotations are specified. This tutorial will demonstate the example of the inputs and showcase the main features of this function.</p>
<section class="level2"><h2 id="prepare-setup">Prepare setup<a class="anchor" aria-label="anchor" href="#prepare-setup"></a>
</h2>
<p>We will first import the necessary packages:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<p>Next, we will get some data to display. The metadata is expected to be a data frame with one required column: <code>sample_id</code> and another column that will contain sample annotations according to the comparison group. In this example, we will use as example the data set and variant calls from the <a href="https://doi.org/10.1182/blood.2022016534" class="external-link">study</a> that identified genetic subgroup of Burkitt lymphoma (BL).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">study</span> <span class="op">==</span> <span class="st">"BL_Thomas"</span><span class="op">)</span> </span></code></pre></div>
</div>
<p>Next, with the help of GAMBLR.open and GAMBLR.data, we will obtain the coding mutations that will be used in the plotting. Many of the <code>get_</code> family of functions return a data frame with mutations annotated in the familiar MAF format.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_coding_ssm.html" class="external-link">get_coding_ssm</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    tool_name <span class="op">=</span> <span class="st">"publication"</span>,</span>
<span>    projection <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>    include_silent <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># What does it look like?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">maf</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 34453    49</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">maf</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>        <span class="va">Tumor_Sample_Barcode</span>,</span>
<span>        <span class="va">Hugo_Symbol</span>,</span>
<span>        <span class="va">Variant_Classification</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>genomic_data Object
Genome Build: hg38
Showing first 10 rows:
  Tumor_Sample_Barcode Hugo_Symbol Variant_Classification
1                Akata        CPTP      Missense_Mutation
2                Akata      FNDC10      Missense_Mutation
3                Akata       MORN1      Missense_Mutation
4                Akata       MEGF6      Missense_Mutation
5                Akata      GPR157      Missense_Mutation
6                Akata        H6PD      Missense_Mutation</code></pre>
</div>
</div>
<p>For the purpose of this tutorial, we will focus on a small subset of genes known to be significantly mutated in BL.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#genes &lt;- lymphoma_genes_bl_v_latest$Gene</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GAMBLR.data/man/lymphoma_genes_lymphoma_genes_v0.0.html" class="external-link">lymphoma_genes</a></span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">BL_Tier</span> <span class="op">==</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ARID1A" "BACH2"  "BCL6"   "BCL7A"  "BMP7"   "CCND3" </code></pre>
</div>
</div>
<p>Now we have our metadata and mutations we want to explore, so we are ready to start visualizing the data.</p>
</section><section class="level2"><h2 id="the-default-forest-plot">The default forest plot<a class="anchor" aria-label="anchor" href="#the-default-forest-plot"></a>
</h2>
<p>The forest plot is ready to be called with the default parameters after just providing the metadata and data frame with mutations in standard maf format. Here is an example of the output with all default parameters:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comparison_column</span> <span class="op">&lt;-</span> <span class="st">"EBV_status_inf"</span> <span class="co"># character of column name comparison</span></span>
<span><span class="va">comparison_vals</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EBV-positive"</span>,<span class="st">"EBV-negative"</span><span class="op">)</span></span>
<span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyForestPlot.html">prettyForestPlot</a></span><span class="op">(</span></span>
<span>    metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    comparison_column <span class="op">=</span> <span class="va">comparison_column</span>,</span>
<span>    comparison_values <span class="op">=</span> <span class="va">comparison_vals</span></span>
<span><span class="op">)</span> </span></code></pre></div>
</div>
<p>The output of the function is a list containing the following objects: - <code>fisher</code>: a data frame with detailed statistics of the Fisherâ€™s test for each gene - <code>mutmat</code>: a binary matrix used for the Fisherâ€™s test - <code>forest</code>: a ggplot2 object with the forest plot of the ORs from the Fisherâ€™s test for each gene - <code>bar</code>: a ggplot2 object wiht mutation frequencies for each Gene - <code>arranged</code>: a display item where both the forest and bar plots are nicely arranged side-by-side</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">fp</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "fisher"   "forest"   "bar"      "arranged" "mutmat"  </code></pre>
</div>
</div>
</section><section class="level2"><h2 id="report-only-significant-differences">Report only significant differences<a class="anchor" aria-label="anchor" href="#report-only-significant-differences"></a>
</h2>
<p>By default, all of the genes of interest are reported in the output. After the Fisherâ€™s test is performed, the <code>prettyForestPlot</code> also calculates FDR and we can use it to only report significant differences by providing a significance cutoff with the parameter <code>max_q</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">max_q</span> <span class="op">&lt;-</span> <span class="fl">0.1</span> <span class="co"># only those qith Q value &lt;= 0.1 will be reported</span></span>
<span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyForestPlot.html">prettyForestPlot</a></span><span class="op">(</span></span>
<span>    metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    comparison_column <span class="op">=</span> <span class="va">comparison_column</span>,</span>
<span>    comparison_values <span class="op">=</span> <span class="va">comparison_vals</span>,</span>
<span>    max_q <span class="op">=</span> <span class="va">max_q</span>,</span>
<span>    base_size <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<p>We now can take a look at what genes are passing the significance cutoff:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span><span class="op">$</span><span class="va">arranged</span> </span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="forestplot_files/figure-html/fdr_plot-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<p>If we donâ€™t provide a set of candidate genes, the Fisherâ€™s exact test is applied to every gene mutated in at least one sample, which is incredibly ambitious. Because <code>prettyForestPlot</code> automatically corrests for multiple testing, this can limit how many significant associations you end up with. Here, we raised the minimum to 10 mutations and excluded the candidate gene restriction. As you can see, this can lead to additional differences being detected.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyForestPlot.html">prettyForestPlot</a></span><span class="op">(</span></span>
<span>    metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    comparison_column <span class="op">=</span> <span class="va">comparison_column</span>,</span>
<span>    comparison_values <span class="op">=</span> <span class="va">comparison_vals</span>,</span>
<span>    max_q <span class="op">=</span> <span class="va">max_q</span>,</span>
<span>    min_mutations <span class="op">=</span> <span class="va">nmut</span>,</span>
<span>    base_size <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="op">)</span></span>
<span><span class="va">fp</span><span class="op">$</span><span class="va">arranged</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="forestplot_files/figure-html/all_genes-1.png" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">num_genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">fp</span><span class="op">$</span><span class="va">fisher</span><span class="op">)</span></span></code></pre></div>
</div>
<p>A total of 18 remained significant after correcting for multiple tests.</p>
</section><section class="level2"><h2 id="comparing-categories-with-more-than-two-groups">Comparing categories with more than two groups<a class="anchor" aria-label="anchor" href="#comparing-categories-with-more-than-two-groups"></a>
</h2>
<p>As the <code>prettyForestPlot</code> construcst the 2x2 contingency tables to run Fisherâ€™s test to find significant differences, it can only operate on comparing 2 groups between themselves - but what if you have more than that and want to see the difference between some of them? To handle this scenario, we can take advantage of the <code>comparison_values</code> parameter, which will be used to subset the metadata to only requested groups and only perform testing and plotting on this subset. Letâ€™s see it in action:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comparison_column</span> <span class="op">&lt;-</span> <span class="st">"genetic_subgroup"</span> <span class="co"># change the comparison column</span></span>
<span><span class="va">comparison_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"IC-BL"</span>, <span class="st">"Q53-BL"</span><span class="op">)</span></span>
<span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyForestPlot.html">prettyForestPlot</a></span><span class="op">(</span></span>
<span>    metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    comparison_column <span class="op">=</span> <span class="va">comparison_column</span>,</span>
<span>    comparison_values <span class="op">=</span> <span class="va">comparison_values</span>,</span>
<span>    max_q <span class="op">=</span> <span class="va">max_q</span>,</span>
<span>    base_size <span class="op">=</span> <span class="fl">7</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">fp</span><span class="op">$</span><span class="va">arranged</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="forestplot_files/figure-html/comp_groups-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<p>This plot is exactly reproducing the Supplemmental Figure 12D from the <a href="https://doi.org/10.1182/blood.2022016534" class="external-link">Thomas et al</a> study!</p>
</section><section class="level2"><h2 id="separating-genes-with-hotspots">Separating genes with hotspots<a class="anchor" aria-label="anchor" href="#separating-genes-with-hotspots"></a>
</h2>
<p>We can additionally separate hotspots from the other mutations and compare those separately. First, we need to annotate the maf data, for which we will use the <code>annotate_hotspots</code> from GAMBLR family. This function will add a new column to the maf named <code>hot_spot</code> indicating whether or not the specific mutation is in the hotspot region.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Annotate hotspots</span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/annotate_hotspots.html" class="external-link">annotate_hotspots</a></span><span class="op">(</span><span class="va">maf</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># What are the hotspots?</span></span>
<span><span class="va">maf</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">hot_spot</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span>, <span class="va">hot_spot</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt; table of extent 0 x 0 &gt;</code></pre>
</div>
</div>
<div>
<blockquote>
<p><strong>Note</strong></p>
<p>The GAMBLR.data version of the <code>annotate_hotspots</code> only handles very specific genes and does not have functionality to annotate all hotspots.</p>
</blockquote>
</div>
<p>Oh no! Looks like there is no hotspots in this maf data. This does not make sense, so what happened? Aha, the hotspot annotation in GAMBLR.data works only on the data in grch37 projection. But our maf is in hg38, so what should we do? One way is to lift the maf data to another projection using the UCSCâ€™s liftOver, and GAMBLR family has exactly the function that serves this purpose:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maf_grch37</span> <span class="op">&lt;-</span> <span class="fu">liftover</span><span class="op">(</span></span>
<span>    <span class="va">maf</span>,</span>
<span>    mode <span class="op">=</span> <span class="st">"maf"</span>,</span>
<span>    target_build <span class="op">=</span> <span class="st">"grch37"</span></span>
<span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Chromosome <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="st">""</span>, <span class="va">Chromosome</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">hot_spot</span><span class="op">)</span> <span class="co"># since it is empty we can just drop it</span></span></code></pre></div>
</div>
<p>Can we annotate the hotspots now?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maf_grch37</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/annotate_hotspots.html" class="external-link">annotate_hotspots</a></span><span class="op">(</span><span class="va">maf_grch37</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># What are the hotspots?</span></span>
<span><span class="va">maf_grch37</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">hot_spot</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span>, <span class="va">hot_spot</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           hot_spot
Hugo_Symbol TRUE
     CREBBP    1
     EZH2      1
     FOXO1    58
     MYD88     2
     STAT6     4</code></pre>
</div>
</div>
<p>Indeed, the hotspots are properly annotated once we have maf in correct projection. Now, we can simply toggle the <code>separate_hotspots</code> parameter to perform separate comparisons within hotspots:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comparison_column</span> <span class="op">&lt;-</span> <span class="st">"EBV_status_inf"</span></span>
<span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyForestPlot.html">prettyForestPlot</a></span><span class="op">(</span></span>
<span>    metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf <span class="op">=</span> <span class="va">maf_grch37</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    comparison_column <span class="op">=</span> <span class="va">comparison_column</span>,</span>
<span>    comparison_values <span class="op">=</span> <span class="va">comparison_vals</span>,</span>
<span>    max_q <span class="op">=</span> <span class="va">max_q</span>,</span>
<span>    separate_hotspots <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">fp</span><span class="op">$</span><span class="va">arranged</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="forestplot_files/figure-html/comp_hotspots-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 id="using-binary-matrix-as-input">Using binary matrix as input<a class="anchor" aria-label="anchor" href="#using-binary-matrix-as-input"></a>
</h2>
<p>Sometimes it might be useful to have different input format instead of maf - for example, you already have a one-hot encoded matrix of feature status (i.e.Â a binary matrix). We can run <code>prettyForestPlot</code> using this as input in lieu of the MAF. This also means we can apply it to more than just simple somatic mutation profiles.</p>
<p>First, letâ€™s construct the binary matrix. We will supplement our maf with the non-coding mutations to look at the aSHM regions in addition to coding mutations, and this will already give us the data in correct projection:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_ssm_by_samples.html" class="external-link">get_ssm_by_samples</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span></span>
<span><span class="op">)</span></span>
<span><span class="va">maf</span><span class="op">$</span><span class="va">Variant_Classification</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="va">table</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>.
               3'Flank                  3'UTR                5'Flank
                  1457                    513                   2957
                 5'UTR        Frame_Shift_Del        Frame_Shift_Ins
                  1102                    124                     97
                   IGR           In_Frame_Del           In_Frame_Ins
                   457                     40                     14
                Intron      Missense_Mutation      Nonsense_Mutation
                 44397                   1859                    286
      Nonstop_Mutation                    RNA                 Silent
                     6                     74                    481
         Splice_Region            Splice_Site Translation_Start_Site
                   148                    114                     22 </code></pre>
</div>
</div>
<p>Now we convert this maf into binary matrix:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Generate binary matrix</span></span>
<span><span class="va">coding_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_coding_ssm_status.html" class="external-link">get_coding_ssm_status</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_data <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    gene_symbols <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    include_hotspots <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    review_hotspots <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<p>Next, supplement this with the matrix of non-coding mutation across aSHM regions</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We'll use the aSHM regions defined in GAMBLR.data</span></span>
<span><span class="co"># Let our helper function massage it into a consistent format</span></span>
<span><span class="va">regions_bed</span> <span class="op">=</span> <span class="fu">create_bed_data</span><span class="op">(</span><span class="va">somatic_hypermutation_locations_GRCh37_v0.2</span>,</span>
<span>                              fix_names<span class="op">=</span><span class="st">"concat"</span>,</span>
<span>                              concat_cols<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>,<span class="st">"region"</span><span class="op">)</span>,</span>
<span>                              sep<span class="op">=</span><span class="st">"-"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generate matrix of mutations per each site</span></span>
<span><span class="va">ashm_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_ashm_count_matrix.html" class="external-link">get_ashm_count_matrix</a></span><span class="op">(</span></span>
<span>    regions_bed <span class="op">=</span> <span class="va">regions_bed</span>,</span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Binarize matrix using arbitrary 3 muts/region cutoff</span></span>
<span><span class="va">ashm_matrix</span><span class="op">[</span><span class="va">ashm_matrix</span> <span class="op">&lt;=</span> <span class="fl">3</span><span class="op">]</span> <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="va">ashm_matrix</span><span class="op">[</span><span class="va">ashm_matrix</span> <span class="op">&gt;</span> <span class="fl">3</span><span class="op">]</span> <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="va">ashm_matrix</span> <span class="op">&lt;-</span> <span class="va">ashm_matrix</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">)</span></span></code></pre></div>
</div>
<p>We can now combine both coding and non-coding features into single matrix:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">feature_matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>    <span class="va">coding_matrix</span>,</span>
<span>    <span class="va">ashm_matrix</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Drop any fearures absent across at least 10 samples to clean any noise</span></span>
<span><span class="va">feature_matrix</span> <span class="op">&lt;-</span> <span class="va">feature_matrix</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select_all.html" class="external-link">select_if</a></span><span class="op">(</span><span class="va">is.numeric</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/where.html" class="external-link">where</a></span><span class="op">(</span><span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">.</span> <span class="op">&gt;</span> <span class="fl">0</span>, na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html" class="external-link">bind_cols</a></span><span class="op">(</span></span>
<span>        <span class="va">feature_matrix</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">sample_id</span><span class="op">)</span>,</span>
<span>        <span class="va">.</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<p>Now we can provide the binary matrix to the <code>prettyForestPlot</code> and regenerate the Supplemmental Figure 12C from the <a href="https://doi.org/10.1182/blood.2022016534" class="external-link">Thomas et al</a> study!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comparison_column</span> <span class="op">&lt;-</span> <span class="st">"genetic_subgroup"</span></span>
<span><span class="va">comparison_values</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DGG-BL"</span>, <span class="st">"Q53-BL"</span><span class="op">)</span></span>
<span><span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyForestPlot.html">prettyForestPlot</a></span><span class="op">(</span></span>
<span>    metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    mutmat <span class="op">=</span> <span class="va">feature_matrix</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    comparison_column <span class="op">=</span> <span class="va">comparison_column</span>,</span>
<span>    comparison_values <span class="op">=</span> <span class="va">comparison_values</span>,</span>
<span>    max_q <span class="op">=</span> <span class="va">max_q</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">fp</span><span class="op">$</span><span class="va">arranged</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="forestplot_files/figure-html/comp_mat-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'light-border',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config);
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
