<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Upgrade your Oncoplots • GAMBLR.viz</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="oncoplot_files/libs/quarto-html/popper.min.js"></script><script src="oncoplot_files/libs/quarto-html/tippy.umd.min.js"></script><link href="oncoplot_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="oncoplot_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Upgrade your Oncoplots">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Introduction</h6></li>
    <li><a class="dropdown-item" href="../articles/why.html">GAMBLR.viz</a></li>
    <li><a class="dropdown-item" href="../articles/install.html"></a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/recap_figures.html">Reproducing figures from GAMBL papers</a></li>
    <li><a class="dropdown-item" href="../articles/forestplot.html">Easy forest plots</a></li>
    <li><a class="dropdown-item" href="../articles/CopyNumberHeatmap.html">Visualizing copy number data as heatmaps</a></li>
    <li><a class="dropdown-item" href="../articles/data_q.html">Exploring data quality</a></li>
    <li><a class="dropdown-item" href="../articles/oncoplot.html">Upgrade your Oncoplots</a></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Metadata and GAMBLR colours</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Help</h6></li>
    <li><a class="dropdown-item" href="../articles/faq.html">Frequently Asked Qestions</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Upgrade your Oncoplots</h1>





      <div class="d-none name"><code></code></div>
    </div>






    <p>One of the main features and integral parts of this package is the display of coding mutations across a set of genes in a given cohort. Because it outperforms other tools and generates display items with a lot of supported features and flexibility, it is called <code>prettyOncoplot</code> and it belongs to the <code>pretty</code> family of GAMBLR.viz functions. This tutorial will demonstate how to prepare inputs for it (spoiler alert: no specific formatting is necessary) and what is the format of metadata expected by <code>prettyOncoplot</code> (spoiler alert: just Tumor_Sample_Barcode column and column for any annotation you want to display).</p>
<section class="level2"><h2 id="prepare-setup">Prepare setup<a class="anchor" aria-label="anchor" href="#prepare-setup"></a>
</h2>
<p>We will first import the necessary packages:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<p>Next, we will get some data to display. The metadata is expected to be a data frame with one required column: <code>Tumor_Sample_Barcode</code> and any other optional column that you want to display as annotation track. In this example, we will use as example the data set and variant calls from the <a href="https://doi.org/10.1182/blood.2022018719" class="external-link">study</a> that identified genetic subgroup of Follicular lymphoma (FL) associated with histologic transformation to DLBCL.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">study</span> <span class="op">==</span> <span class="st">"FL_Dreval"</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Next, we will obtain the coding mutations that will be used in the plotting. The data is a data frame in a standartized maf format.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_ssm_by_samples.html" class="external-link">get_ssm_by_samples</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    tool_name <span class="op">=</span> <span class="st">"publication"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># How does it look like?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">maf</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 44777    49</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">maf</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>        <span class="va">Tumor_Sample_Barcode</span>,</span>
<span>        <span class="va">Hugo_Symbol</span>,</span>
<span>        <span class="va">Variant_Classification</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>genomic_data Object
Genome Build: grch37
Showing first 10 rows:
  Tumor_Sample_Barcode Hugo_Symbol Variant_Classification
1             FL1001T1       AIFM1      Nonsense_Mutation
2             FL1001T1       USP24      Missense_Mutation
3             FL1001T1        LYST            Splice_Site
4             FL1001T1        GCKR      Missense_Mutation
5             FL1001T1      ORMDL1      Missense_Mutation
6             FL1001T1        BCL6      Missense_Mutation</code></pre>
</div>
</div>
<div>
<blockquote>
<p><strong>Did you know?</strong></p>
<p>You do not have to subset your maf data frame to coding mutations only before using it with the <code>prettyOncoplot</code>. Much like other tools, it will be automatically handled for you to only display coding mutations.</p>
</blockquote>
</div>
<p>Now we have our metadata and mutations we want to explore, so we are ready to start visualizing the data.</p>
</section><section class="level2"><h2 id="the-simplest-oncoplot">The simplest oncoplot<a class="anchor" aria-label="anchor" href="#the-simplest-oncoplot"></a>
</h2>
<p>There is a number of options how to customize your oncoplot, but it is ready for you to use with just the metadata and maf. Here is an example of the output with all default parameters:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">minMutationPercent</span> <span class="op">&lt;-</span> <span class="fl">10</span> <span class="co"># only show genes mutated in at least 10% of samples</span></span>
<span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    minMutationPercent <span class="op">=</span> <span class="va">minMutationPercent</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="oncoplot_files/figure-html/default-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 id="adding-annotation-tracks">Adding annotation tracks<a class="anchor" aria-label="anchor" href="#adding-annotation-tracks"></a>
</h2>
<p>We can customize this and add some of the annotation tracks for more informative display of the metadata we ate interested in:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadataColumns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"pathology"</span>,</span>
<span>    <span class="st">"lymphgen"</span>,</span>
<span>    <span class="st">"genetic_subgroup"</span>,</span>
<span>    <span class="st">"COO_consensus"</span>,</span>
<span>    <span class="st">"sex"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    minMutationPercent <span class="op">=</span> <span class="va">minMutationPercent</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="oncoplot_files/figure-html/add_annotations-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 id="changing-font-sizes">Changing font sizes<a class="anchor" aria-label="anchor" href="#changing-font-sizes"></a>
</h2>
<p>You may notice that as more (or less) genes and annotations are displayed with the oncoplot we may want to modify the size of the gene names and/or the annotation tracks with their labels. There are several parameters available for you to do so: - <code>metadataBarHeight</code>: will change the height of the annotation tracks at the bottom of the oncoplot - <code>metadataBarFontsize</code>: will change the font size of the annotation tracks at the bottom of the oncoplot - <code>fontSizeGene</code>: will change the font size of both percentage labels to the right of the oncoplot and gene names to the left of it - <code>legendFontSize</code>: will change the font size of the legend at the bottom of the plot Let’s see these parameters in action:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadataBarHeight</span> <span class="op">&lt;-</span> <span class="fl">5</span></span>
<span><span class="va">metadataBarFontsize</span> <span class="op">&lt;-</span> <span class="fl">10</span></span>
<span><span class="va">fontSizeGene</span> <span class="op">&lt;-</span> <span class="fl">12</span></span>
<span><span class="va">legendFontSize</span> <span class="op">&lt;-</span> <span class="fl">7</span></span>
<span></span>
<span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    minMutationPercent <span class="op">=</span> <span class="va">minMutationPercent</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="oncoplot_files/figure-html/adjust_fonts-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 id="show-samples-ordered-on-annotations">Show samples ordered on annotations<a class="anchor" aria-label="anchor" href="#show-samples-ordered-on-annotations"></a>
</h2>
<p>We can notice that the default setting generates the classic “rainfall” style of the plot - but what if we want to add some structure to it and sort sample order in some way? It is easy to do so with the parameter <code>sortByColumns</code>. We can sort on the same annotations as we use to display with the oncoplot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    minMutationPercent <span class="op">=</span> <span class="va">minMutationPercent</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="va">metadataColumns</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="oncoplot_files/figure-html/sort_samples-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<div>
<blockquote>
<p><strong>Did you know?</strong></p>
<p>The ordering occurs sequentially according to the order of individual columns we have specified with the <code>sortByColumns</code> parameter. The ordering is in ascending order, and can be toggled with additional boolean parameter <code>arrange_descending</code>.</p>
</blockquote>
</div>
</section><section class="level2"><h2 id="displaying-only-specific-genes">Displaying only specific genes<a class="anchor" aria-label="anchor" href="#displaying-only-specific-genes"></a>
</h2>
<p>There can be scenarion where we might want to diplay genes not based on their recurrence, but out of interest in specific genes. Sure so, one way to do it is to pre-filter your maf data to the genes of interest. But this might have some unexpected consequences and limit your flexibility in doing more things, so the better way is to take advantage of the <code>genes</code> parameter:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fl_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"RRAGC"</span>, <span class="st">"CREBBP"</span>, <span class="st">"VMA21"</span>, <span class="st">"ATP6V1B2"</span>, <span class="st">"EZH2"</span>, <span class="st">"KMT2D"</span><span class="op">)</span></span>
<span><span class="va">dlbcl_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MEF2B"</span>, <span class="st">"CD79B"</span>, <span class="st">"MYD88"</span>, <span class="st">"TP53"</span><span class="op">)</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">fl_genes</span>, <span class="va">dlbcl_genes</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="oncoplot_files/figure-html/goi-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<div>
<blockquote>
<p><strong>Note</strong></p>
<p>Note that we removed the <code>minMutationPercent</code> in the last function call since we wanted to see the genes that we specifically requested.</p>
</blockquote>
</div>
<p>Now we are only looking at some specific genes of interest but they are arranged in the decreasing order of their recurrence in this cohort. What if we want to enforce the gene order on the oncoplot to be exactly the same as we specified it in our <code>gene</code> variable? We can take advantage of the <code>keepGeneOrder</code> parameter:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    keepGeneOrder <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="oncoplot_files/figure-html/goi_order-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 id="grouping-genes-into-categories">Grouping genes into categories<a class="anchor" aria-label="anchor" href="#grouping-genes-into-categories"></a>
</h2>
<p>We can also group genes into specific categories. To do so, we need to have a named list where name of the list element corresponds to the gene name, and the list element corresponds to the gene group. We alreade have the <code>genes</code> variable, so we can convert it to the appropriate format:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gene_groups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"FL"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">fl_genes</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">"DLBCL"</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">dlbcl_genes</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">gene_groups</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">genes</span></span>
<span></span>
<span><span class="va">gene_groups</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   RRAGC   CREBBP    VMA21 ATP6V1B2     EZH2    KMT2D    MEF2B    CD79B
    "FL"     "FL"     "FL"     "FL"     "FL"     "FL"  "DLBCL"  "DLBCL"
   MYD88     TP53
 "DLBCL"  "DLBCL" </code></pre>
</div>
</div>
<p>Now we can use it to split the genes on the oncoplot into the groups:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="oncoplot_files/figure-html/goi_groups-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<div>
<blockquote>
<p><strong>Did you know?</strong></p>
<p>You can provide more than two groups of genes - any number of groups is supported as long as they are specified in the <code>gene_groups</code>.</p>
</blockquote>
</div>
<div>
<blockquote>
<p><strong>Note</strong></p>
<p>Within each group, the genes are ordered in decreasing order of their recurrence, but the <code>keepGeneOrder</code> parameter is still supported and if specified, will keep the specified order within each group.</p>
</blockquote>
</div>
</section><section class="level2"><h2 id="grouping-samples-into-categories">Grouping samples into categories<a class="anchor" aria-label="anchor" href="#grouping-samples-into-categories"></a>
</h2>
<p>Similar to the grouping of genes, we can also group samples into certain categories. Typically, it is done based on one of the annotations tracks. By default, there will be no labels for each sample category, but we also have an option of specifying these labels:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span>,</span>
<span>    splitColumnName <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>    groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Follicular lymphoma"</span>, <span class="st">"DLBCL"</span>, <span class="st">"COMFL"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="oncoplot_files/figure-html/sample_groups-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 id="tallying-mutation-burden">Tallying mutation burden<a class="anchor" aria-label="anchor" href="#tallying-mutation-burden"></a>
</h2>
<p>Previously, we noted that the maf data we were supplying to the prettyOncoplot was not subset to contain only coding mutations, and also discouraged from pre-filtering maf to a subset of genes if we are insterested only looking at some of them. <strong>Here is why this is important:</strong> if we want to layer on additional information like total mutation burden per sample, any subsetting or filtering of the maf would generate inaccurate and misleading results. Therefore, <code>prettyOncoplot</code> handles all of this for you! So if we were to go ahead with tallying the total mutation burden, we could just add some additional parameters to the function call:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hideTopBarplot</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span> <span class="co"># will display TMB annotations at the top</span></span>
<span><span class="va">tally_all_mutations</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span> <span class="co"># will tally all mutations per sample</span></span>
<span></span>
<span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span>,</span>
<span>    splitColumnName <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>    groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Follicular lymphoma"</span>, <span class="st">"DLBCL"</span>, <span class="st">"COMFL"</span><span class="op">)</span>,</span>
<span>    hideTopBarplot <span class="op">=</span> <span class="va">hideTopBarplot</span>,</span>
<span>    tally_all_mutations <span class="op">=</span> <span class="va">tally_all_mutations</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="oncoplot_files/figure-html/tmb-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<div>
<blockquote>
<p><strong>Did you know?</strong></p>
<p>If the dynamic range of total mutation burden is too big and there are some extreme outliers, the bar chart at the top of the oncoplot can be capped of at any numeric value by providing <code>tally_all_mutations_max</code> parameter.</p>
</blockquote>
</div>
<p>What if we want to additionally force the ordering based on the total number of mutations, so they are nicely arranged in the decreasing order? We can do so by adding the mutation counts as one of the annotation tracks and using it to sort the samples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Count all muts to define the order of samples</span></span>
<span><span class="va">total_mut_burden</span> <span class="op">&lt;-</span> <span class="va">maf</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">Tumor_Sample_Barcode</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">total_mut_burden</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>genomic_data Object
Genome Build: grch37
Showing first 10 rows:
  Tumor_Sample_Barcode   n
1            01-20260T  71
2            02-13135T  98
3            02-20170T  67
4            02-22991T  53
5            03-34157T  26
6            04-24937T 146</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add this info to metadata</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>    <span class="va">metadata</span>,</span>
<span>    <span class="va">total_mut_burden</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n"</span>, <span class="va">metadataColumns</span><span class="op">)</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span>,</span>
<span>    splitColumnName <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>    groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Follicular lymphoma"</span>, <span class="st">"DLBCL"</span>, <span class="st">"COMFL"</span><span class="op">)</span>,</span>
<span>    hideTopBarplot <span class="op">=</span> <span class="va">hideTopBarplot</span>,</span>
<span>    tally_all_mutations <span class="op">=</span> <span class="va">tally_all_mutations</span>,</span>
<span>    numericMetadataColumns <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>    arrange_descending <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="oncoplot_files/figure-html/tmb_order_by_meta-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<div>
<blockquote>
<p><strong>Note</strong></p>
<p>We have modified here the <code>sortByColumns</code> parameter, and provided two additional parameters <code>numericMetadataColumns</code> and <code>arrange_descending</code>.</p>
</blockquote>
</div>
<div>
<blockquote>
<p><strong>Did you know?</strong></p>
<p>The top annotation and <code>n</code> annotation at the bottom are the same thing? Remove <code>n</code> from the legend by adding <code>hide_annotations = "n"</code> and remove display of annotation track while keeping the ordering by adding <code>hide_annotations_tracks = TRUE</code>.</p>
</blockquote>
</div>
</section><section class="level2"><h2 id="annotating-significance-of-mutation-frequencies-in-sample-groups">Annotating significance of mutation frequencies in sample groups<a class="anchor" aria-label="anchor" href="#annotating-significance-of-mutation-frequencies-in-sample-groups"></a>
</h2>
<p>When looking at our sample plots, we can notice that the frequency of mutations in <em>RRAGC</em>, <em>ATP6V1B2</em>, <em>VMA21</em> and others is different between FL and DLBCL. But is this difference significant? Can we layer on this diffenerence to the display panel? Yes we can, and this is very easy with GAMBLR family! To do so we will first use another function from GAMBLR.viz to run Fisher’s test and find which genes are significantly different between the FL and DLBCL:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fisher_test</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyForestPlot.html">prettyForestPlot</a></span><span class="op">(</span></span>
<span>    maf <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    comparison_column <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>    comparison_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCL"</span>, <span class="st">"FL"</span><span class="op">)</span>, <span class="co"># we have three pathologies in data</span></span>
<span>    comparison_name <span class="op">=</span> <span class="st">"FL vs DLBCL"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">fisher_test</span><span class="op">$</span><span class="va">arranged</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="oncoplot_files/figure-html/fisher-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
<p>In fact, there are genes that are mutated at significantly different frequencies! Now let’s layer on this information to our oncoplot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n"</span>, <span class="va">metadataColumns</span><span class="op">)</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span>,</span>
<span>    splitColumnName <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>    groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Follicular lymphoma"</span>, <span class="st">"DLBCL"</span>, <span class="st">"COMFL"</span><span class="op">)</span>,</span>
<span>    hideTopBarplot <span class="op">=</span> <span class="va">hideTopBarplot</span>,</span>
<span>    tally_all_mutations <span class="op">=</span> <span class="va">tally_all_mutations</span>,</span>
<span>    numericMetadataColumns <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>    arrange_descending <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    hide_annotations <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>    hide_annotations_tracks <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    annotate_specific_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    this_forest_object <span class="op">=</span> <span class="va">fisher_test</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level2"><h2 id="annotating-genes-with-hotspots">Annotating genes with hotspots<a class="anchor" aria-label="anchor" href="#annotating-genes-with-hotspots"></a>
</h2>
<p>Some genes are mutated at certain positions more often that at others, therefore creating the mutational hotspots - and it we can layer on this level of information to our oncoplot. First, we will need to process our maf data to add a new column called <code>hot_spot</code> which will contain a boolean value showing whether or not particular mutation is a hotspot. If you don’t know how to do it, there is a function for exactly this purpose in the <a href="https://github.com/morinlab/GAMBLR.data" class="external-link">GAMBLR.data</a>, and we will use it in this example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Annotate hotspots</span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/annotate_hotspots.html" class="external-link">annotate_hotspots</a></span><span class="op">(</span><span class="va">maf</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># What are the hotspots?</span></span>
<span><span class="va">maf</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">hot_spot</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">Hugo_Symbol</span>, <span class="va">hot_spot</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           hot_spot
Hugo_Symbol TRUE
     CREBBP   76
     EZH2     86
     FOXO1    20
     MEF2B    23
     MYD88    46
     STAT6    51</code></pre>
</div>
</div>
<div>
<blockquote>
<p><strong>Note</strong></p>
<p>The GAMBLR.data version of the <code>annotate_hotspots</code> only handles very specific genes and does not have functionality to annotate all hotspots.</p>
</blockquote>
</div>
<p>Now, we can add annotation of the hotspots to the oncoplot display by toggling the <code>highlightHotspots</code> parameter:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">highlightHotspots</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span>
<span><span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n"</span>, <span class="va">metadataColumns</span><span class="op">)</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span>,</span>
<span>    splitColumnName <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>    groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Follicular lymphoma"</span>, <span class="st">"DLBCL"</span>, <span class="st">"COMFL"</span><span class="op">)</span>,</span>
<span>    hideTopBarplot <span class="op">=</span> <span class="va">hideTopBarplot</span>,</span>
<span>    tally_all_mutations <span class="op">=</span> <span class="va">tally_all_mutations</span>,</span>
<span>    numericMetadataColumns <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>    arrange_descending <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    hide_annotations <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>    hide_annotations_tracks <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    annotate_specific_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    this_forest_object <span class="op">=</span> <span class="va">fisher_test</span>,</span>
<span>    highlightHotspots <span class="op">=</span> <span class="va">highlightHotspots</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level2"><h2 id="co-oncoplot-two-plots-side-by-side">Co-oncoplot: two plots side-by-side<a class="anchor" aria-label="anchor" href="#co-oncoplot-two-plots-side-by-side"></a>
</h2>
<p>It may also be informative to generate a display panel where there are two oncoplots displayed side-by-side, so it is possible to visually compare the specific groups of samples while maintaining all annotations and ordering we built so far. For this purpose, the GAMBLR.viz has another function in the <code>pretty</code> family: <code>prettyCoOncoplot</code>. It accepts all of the same parameters as <code>prettyOncoplot</code> with addition of some unique additions. For example, lets break down our sample oncoplot we created so far by the <code>genetic_subgroup</code> and see how cFL compares to dFL:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/prettyCoOncoplot.html">prettyCoOncoplot</a></span><span class="op">(</span></span>
<span>    metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    comparison_column <span class="op">=</span> <span class="st">"genetic_subgroup"</span>,</span>
<span>    label1 <span class="op">=</span> <span class="st">"cFL"</span>,</span>
<span>    label2 <span class="op">=</span> <span class="st">"dFL"</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n"</span>, <span class="va">metadataColumns</span><span class="op">)</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span>,</span>
<span>    splitColumnName <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>    keepGeneOrder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Follicular lymphoma"</span>, <span class="st">"DLBCL"</span>, <span class="st">"COMFL"</span><span class="op">)</span>,</span>
<span>    hideTopBarplot <span class="op">=</span> <span class="va">hideTopBarplot</span>,</span>
<span>    tally_all_mutations <span class="op">=</span> <span class="va">tally_all_mutations</span>,</span>
<span>    numericMetadataColumns <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>    arrange_descending <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    hide_annotations <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>    hide_annotations_tracks <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    annotate_specific_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    this_forest_object <span class="op">=</span> <span class="va">fisher_test</span>,</span>
<span>    highlightHotspots <span class="op">=</span> <span class="va">highlightHotspots</span>,</span>
<span>    legend_row <span class="op">=</span> <span class="fl">2</span>,</span>
<span>    annotation_row <span class="op">=</span> <span class="fl">2</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div>
<blockquote>
<p><strong>Note</strong></p>
<p>It is only possible to display two groups side-by-side. If the metadata column you want to split on contains more groups, the specific values can be specified with <code>comparison_values</code> parameter.</p>
</blockquote>
</div>
<div>
<blockquote>
<p><strong>Did you know?</strong></p>
<p>Notice that we did not need to create individual maf or metadata objects to supply to <code>prettyCoOncoplot</code> - the same objects we used before are also supported here, but specified with differen parameters <code>metadata</code> and <code>maf</code>.</p>
</blockquote>
</div>
<p>In the above example, we forced the order of genes to be exaclty as we specified so that the same gene is is displayed on the same row for both oncoplots, othervise they wold not be on the same row due to the different frequencies in each group. In addition to specifying this parameter, we have also enforced specific number of rows in the legend below the plot, so they nicely align between the display items.</p>
</section><section class="level2"><h2 id="using-oncoplot-in-multi-panel-figure">Using oncoplot in multi-panel figure<a class="anchor" aria-label="anchor" href="#using-oncoplot-in-multi-panel-figure"></a>
</h2>
<p>When arranging items for the multi-panel figure when preparing manuscript or experiment report, it may be needed to use the generated oncoplot on the same page as other display items. The <code>prettyOncoplot</code> (and, therefore, <code>prettyCoOncoplot</code>), handles the ComplexHeatmap under the hood to generate graphics, and it is not readily available to be combined with the plots generated with other tools, for example <code>ggplot2</code>. Not readily available - but definitely not impossible! The output of <code>prettyCoOncoplot</code> is directly compatible with the arrangement on multi-panel figure since it uses the trick shown below under the hood to put two panels side-by-side, but the otuput of <code>prettyOncoplot</code> is a ComplexHeatmap object so needs some extra steps to allow multi-panel arrangement. First, lets store the returned oncoplot in a variable:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_oncoplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n"</span>, <span class="va">metadataColumns</span><span class="op">)</span>,</span>
<span>    genes <span class="op">=</span> <span class="va">genes</span>,</span>
<span>    splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span>,</span>
<span>    splitColumnName <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>    groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Follicular lymphoma"</span>, <span class="st">"DLBCL"</span>, <span class="st">"COMFL"</span><span class="op">)</span>,</span>
<span>    hideTopBarplot <span class="op">=</span> <span class="va">hideTopBarplot</span>,</span>
<span>    tally_all_mutations <span class="op">=</span> <span class="va">tally_all_mutations</span>,</span>
<span>    numericMetadataColumns <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>    arrange_descending <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    hide_annotations <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>    hide_annotations_tracks <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    annotate_specific_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    this_forest_object <span class="op">=</span> <span class="va">fisher_test</span>,</span>
<span>    highlightHotspots <span class="op">=</span> <span class="va">highlightHotspots</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<p>Next, we will import some of the packages needed to handle the trick:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span> <span class="co"># to handle the ComplexHeatmap object</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span> <span class="co"># to arrange multiple panels</span></span></code></pre></div>
</div>
<p>After that, we will capture the display of the oncoplot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_oncoplot</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/grid.grab.html" class="external-link">grid.grabExpr</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">my_oncoplot</span><span class="op">)</span>,</span>
<span>    width <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    height <span class="op">=</span> <span class="fl">17</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<p>Now, it is ready for us to arrange in multi-panel figure. We can use the forest plot we already looked at as an example, and put it to the right of the oncoplot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multipanel_figure</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span></span>
<span>    <span class="va">my_oncoplot</span>, <span class="co"># left panel</span></span>
<span>    <span class="va">fisher_test</span><span class="op">$</span><span class="va">arranged</span>, <span class="co"># right panel</span></span>
<span>    widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">1</span><span class="op">)</span>, <span class="co"># so the oncoplot is a little wider than the forest</span></span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="co"># labels for the panels</span></span>
<span>    font.label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span> <span class="co"># make labels bold face</span></span>
<span>        color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>        face <span class="op">=</span> <span class="st">"bold"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">multipanel_figure</span></span></code></pre></div>
</div>
<p>Final note: it would be nice to have the genes in the forest plot directly aligned with the genes as they are displayed on the oncoplot, and we can do this by providing consistent ordering and adding some white space below forest plot to match the height of the oncoplot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">my_oncoplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyOncoplot.html">prettyOncoplot</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_df <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="va">metadataColumns</span>,</span>
<span>    metadataBarHeight <span class="op">=</span> <span class="va">metadataBarHeight</span>,</span>
<span>    metadataBarFontsize <span class="op">=</span> <span class="va">metadataBarFontsize</span>,</span>
<span>    fontSizeGene <span class="op">=</span> <span class="va">fontSizeGene</span>,</span>
<span>    legendFontSize <span class="op">=</span> <span class="va">legendFontSize</span>,</span>
<span>    sortByColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n"</span>, <span class="va">metadataColumns</span><span class="op">)</span>,</span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="va">fisher_test</span><span class="op">$</span><span class="va">fisher</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span>,</span>
<span>    keepGeneOrder <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    splitGeneGroups <span class="op">=</span> <span class="va">gene_groups</span>,</span>
<span>    splitColumnName <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>    groupNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Follicular lymphoma"</span>, <span class="st">"DLBCL"</span>, <span class="st">"COMFL"</span><span class="op">)</span>,</span>
<span>    hideTopBarplot <span class="op">=</span> <span class="va">hideTopBarplot</span>,</span>
<span>    tally_all_mutations <span class="op">=</span> <span class="va">tally_all_mutations</span>,</span>
<span>    numericMetadataColumns <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>    arrange_descending <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    hide_annotations <span class="op">=</span> <span class="st">"n"</span>,</span>
<span>    hide_annotations_tracks <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    annotate_specific_genes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    this_forest_object <span class="op">=</span> <span class="va">fisher_test</span>,</span>
<span>    highlightHotspots <span class="op">=</span> <span class="va">highlightHotspots</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">my_oncoplot</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/grid.grab.html" class="external-link">grid.grabExpr</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span><span class="va">my_oncoplot</span><span class="op">)</span>,</span>
<span>    width <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    height <span class="op">=</span> <span class="fl">17</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">multipanel_figure</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span></span>
<span>    <span class="va">my_oncoplot</span>, <span class="co"># left panel</span></span>
<span>    <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span> <span class="co"># right panel</span></span>
<span>        <span class="cn">NULL</span>, <span class="co"># empty space at the top</span></span>
<span>        <span class="va">fisher_test</span><span class="op">$</span><span class="va">arranged</span>, <span class="co"># forest on the top</span></span>
<span>        <span class="cn">NULL</span>, <span class="co"># empty space at the bottom</span></span>
<span>        nrow <span class="op">=</span> <span class="fl">3</span>, <span class="co"># arrange vertically</span></span>
<span>        heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.1</span>, <span class="fl">2.5</span>, <span class="fl">1</span><span class="op">)</span> <span class="co"># match height of the oncoplot</span></span>
<span>    <span class="op">)</span>,</span>
<span>    widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">1</span><span class="op">)</span>, <span class="co"># so the oncoplot is a little wider than the forest</span></span>
<span>    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="co"># labels for the panels</span></span>
<span>    font.label <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span> <span class="co"># make labels bold face</span></span>
<span>        color <span class="op">=</span> <span class="st">"black"</span>,</span>
<span>        face <span class="op">=</span> <span class="st">"bold"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">multipanel_figure</span></span></code></pre></div>
</div>
</section><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'light-border',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config);
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
