<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Reproducing figures from GAMBL papers â€¢ GAMBLR.viz</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><script src="recap_figures_files/libs/quarto-html/popper.min.js"></script><script src="recap_figures_files/libs/quarto-html/tippy.umd.min.js"></script><link href="recap_figures_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="recap_figures_files/libs/quarto-html/light-border.css" rel="stylesheet">
<!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Reproducing figures from GAMBL papers">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-viz.html">Get started</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Installation</h6></li>
    <li><a class="dropdown-item" href="../articles/install.html">Installation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/recap_figures.html">Reproducing figures from GAMBL papers</a></li>
    <li><a class="dropdown-item" href="../articles/forestplot.html">Easy forest plots</a></li>
    <li><a class="dropdown-item" href="../articles/CopyNumberHeatmap.html">Visualizing copy number data as heatmaps</a></li>
    <li><a class="dropdown-item" href="../articles/data_q.html">Exploring data quality</a></li>
    <li><a class="dropdown-item" href="../articles/oncoplot.html">Upgrade your Oncoplots</a></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Metadata and GAMBLR colours</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Help</h6></li>
    <li><a class="dropdown-item" href="../articles/faq.html">Frequently Asked Questions</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-quarto">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Reproducing figures from GAMBL papers</h1>





      <div class="d-none name"><code></code></div>
    </div>






    <p>When preparing a manuscript for publication, it is important to maintain a visual identity throughout display items. With GAMBLR.viz, it is easy to keep up with a consistent theme, color pallette, fonts and font sizes. <br> This tutorial will demonstate how to regenerate figures from some of the manuscripts previously published by Morin Lab, in particular study by <a href="https://doi.org/10.1182/blood.2022018719" class="external-link">Dreval et al.</a></p>
<section class="level2"><h2 id="prepare-setup">Prepare setup<a class="anchor" aria-label="anchor" href="#prepare-setup"></a>
</h2>
<p>We will first import the necessary packages:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://readxl.tidyverse.org" class="external-link">readxl</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jokergoo/ComplexHeatmap" class="external-link">ComplexHeatmap</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eclarke/ggbeeswarm" class="external-link">ggbeeswarm</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span></code></pre></div>
</div>
<p>Next, we will use the standard color pallette and store it in a variable for easy access:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu">get_gambl_colours</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Finally, we will set the consistent GAMBLR-specific ggplot theme for this tutorial:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme_get.html" class="external-link">theme_set</a></span><span class="op">(</span><span class="fu">theme_Morons</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</section><section class="level2"><h2 id="differential-gene-expression-between-fl-subgroups">Differential gene expression between FL subgroups<a class="anchor" aria-label="anchor" href="#differential-gene-expression-between-fl-subgroups"></a>
</h2>
<p>The Figure 4A of the <a href="https://doi.org/10.1182/blood.2022018719" class="external-link">Dreval et al</a> shows that there is a differential gene expression of CREBBP, FOXP1, and MYC between genetic subgroups of FL described in that paper. We can regenerate it here to confirm the published results and demonstate how easy it is to work with GAMBLR.viz.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get metadata</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span></span>
<span>        <span class="va">study</span> <span class="op">==</span> <span class="st">"FL_Dreval"</span>,</span>
<span>        <span class="va">pathology</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span>, <span class="st">"DLBCL"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<p>Next, we can take advantage of the publicly available gene expression data and access it directly. The data is provided with the ENSEMBL identifiers, so we will need to make the conversion between human-readable and ENSEMBL identifiers. There is no need to search for this linkage as it is direcly available with GAMBLR family:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">identifiers</span> <span class="op">&lt;-</span> <span class="va">grch37_gene_coordinates</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gene_name</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CREBBP"</span>, <span class="st">"FOXP1"</span>, <span class="st">"MYC"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">ensembl_gene_id</span>, <span class="va">gene_name</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Now we can access the gene expression data and prepare it for plotting</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get gene expression</span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span></span>
<span>        <span class="st">"https://www.bcgsc.ca/downloads/morinlab/cFL_Blood_2023_GE.tsv.gz"</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">ensembl_gene_id</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">identifiers</span><span class="op">$</span><span class="va">ensembl_gene_id</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="co"># add human-readable identifiers</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>        <span class="va">identifiers</span>,</span>
<span>        <span class="va">.</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">expression</span> <span class="op">&lt;-</span> <span class="va">expression</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="co"># convert to long format</span></span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html" class="external-link">pivot_longer</a></span><span class="op">(</span></span>
<span>        <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">ensembl_gene_id</span>, <span class="va">gene_name</span><span class="op">)</span>,</span>
<span>        names_to <span class="op">=</span> <span class="st">"sample_id"</span>,</span>
<span>        values_to <span class="op">=</span> <span class="st">"expression"</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="co"># add genetic metadata labels</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span></span>
<span>        <span class="va">.</span>,</span>
<span>        <span class="va">metadata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>            <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">sample_id</span>, <span class="va">pathology</span>, <span class="va">genetic_subgroup</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="co"># prepare grouping label and set consistent level order</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        comp_group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span></span>
<span>            <span class="va">pathology</span> <span class="op">==</span> <span class="st">"DLBCL"</span>,</span>
<span>            <span class="va">pathology</span>,</span>
<span>            <span class="va">genetic_subgroup</span></span>
<span>        <span class="op">)</span>,</span>
<span>        comp_group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span></span>
<span>            <span class="va">comp_group</span>,</span>
<span>            levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCL"</span>, <span class="st">"dFL"</span>, <span class="st">"cFL"</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/drop_na.html" class="external-link">drop_na</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<p>Now we can perform statistical comparisons to find differences in gene expression:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pwc</span> <span class="op">&lt;-</span> <span class="va">expression</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">gene_name</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">rstatix</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/rstatix/reference/wilcox_test.html" class="external-link">wilcox_test</a></span><span class="op">(</span></span>
<span>        <span class="va">expression</span> <span class="op">~</span> <span class="va">comp_group</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="co"># Add coordinates for position of brackets</span></span>
<span><span class="va">pwc</span> <span class="op">&lt;-</span> <span class="va">pwc</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu">rstatix</span><span class="fu">::</span><span class="fu"><a href="https://rpkgs.datanovia.com/rstatix/reference/get_pvalue_position.html" class="external-link">add_xy_position</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"comp_group"</span><span class="op">)</span></span></code></pre></div>
</div>
<p>The Plot!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">expression</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="va">comp_group</span>,</span>
<span>            y <span class="op">=</span> <span class="va">expression</span>,</span>
<span>            color <span class="op">=</span> <span class="va">comp_group</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_quasirandom.html" class="external-link">geom_quasirandom</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_pvalue_manual.html" class="external-link">stat_pvalue_manual</a></span><span class="op">(</span></span>
<span>        <span class="va">pwc</span>,</span>
<span>        hide.ns <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>        size <span class="op">=</span> <span class="fl">8</span>,</span>
<span>        label <span class="op">=</span> <span class="st">"p.adj.signif"</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_grid.html" class="external-link">facet_grid</a></span><span class="op">(</span>cols <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">gene_name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Expression"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>        axis.title.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span><span class="va">p</span></span></code></pre></div>
</div>
</section><section class="level2"><h2 id="differentially-mutated-genes-between-dlbcl-and-fl">Differentially mutated genes between DLBCL and FL<a class="anchor" aria-label="anchor" href="#differentially-mutated-genes-between-dlbcl-and-fl"></a>
</h2>
<p>The same paper by <a href="https://doi.org/10.1182/blood.2022018719" class="external-link">Dreval et al</a> showed in the Supplemental Figure 1B the genes mutated at differential frequencies between FL and DLBCL. We can also easily regenerate that plot:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First obtain maf data</span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_ssm_by_samples.html" class="external-link">get_ssm_by_samples</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    tool_name <span class="op">=</span> <span class="st">"publication"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The Plot!</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyForestPlot.html">prettyForestPlot</a></span><span class="op">(</span></span>
<span>    maf <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    comparison_column <span class="op">=</span> <span class="st">"pathology"</span>,</span>
<span>    comparison_values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DLBCL"</span>, <span class="st">"FL"</span><span class="op">)</span>,</span>
<span>    max_q <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>    genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">lymphoma_genes</span><span class="op">$</span><span class="va">Gene</span>, <span class="st">"VMA21"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">arranged</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="recap_figures_files/figure-html/fl_forest-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><section class="level2"><h2 id="patterns-of-mutations-at-common-ashm-target-sites">Patterns of mutations at common aSHM target sites<a class="anchor" aria-label="anchor" href="#patterns-of-mutations-at-common-ashm-target-sites"></a>
</h2>
<p>The study of FL by <a href="https://doi.org/10.1182/blood.2022018719" class="external-link">Dreval et al</a> in the Supplemental Figure 4A demonstrated and important visualization of the patterns of mutations at common aSHM target sites across DLBCL and FL when comparing tumors in the discovery cohort. We can recapitulate that figure in this tutorial using the data bundled with GAMBLR and visualization functions available with GAMBLR.viz.</p>
<p>First, we will retreive the metadata exactly as it is provided in the paper:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Read supplemental table from that paper</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_xlsx</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>        <span class="st">"extdata"</span>,</span>
<span>        <span class="st">"studies/FL_Dreval.xlsx"</span>,</span>
<span>        package <span class="op">=</span> <span class="st">"GAMBLR.data"</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># What is provided in the supplemental table?</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">metadata</span><span class="op">)</span></span></code></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Patient barcode"           "Pairing status"
 [3] "Genome sample id"          "Normal sample id"
 [5] "Sex"                       "Age at diagnosis"
 [7] "Pathology"                 "FL grade"
 [9] "Analysis cohort"           "Tumor biopsy"
[11] "Reference"                 "MYC FISH BA"
[13] "BCL2 FISH BA"              "BCL6 FISH BA"
[15] "MYC WGS Tx"                "BCL2 WGS Tx"
[17] "BCL6 WGS Tx"               "Tumor purity"
[19] "total N of SSM"            "PGA"
[21] "Analysis"                  "cFL/dFL label"
[23] "SeqType"                   "AverageBaseQuality"
[25] "AverageInsertSize"         "AverageReadLength"
[27] "PairsOnDiffCHR"            "TotalReads"
[29] "TotalUniquelyMapped"       "TotalUnmappedreads"
[31] "TotalDuplicatedreads"      "ProportionReadsDuplicated"
[33] "ProportionReadsMapped"     "MeanCorrectedCoverage"
[35] "ProportionCoverage10x"     "ProportionCoverage30x"    </code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Select and rename columns that we need for heatmap</span></span>
<span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span></span>
<span>        sample_id <span class="op">=</span> <span class="va">`Genome sample id`</span>,</span>
<span>        <span class="va">`Analysis cohort`</span>,</span>
<span>        <span class="st">"BCL2 Status"</span> <span class="op">=</span> <span class="va">`BCL2 WGS Tx`</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        <span class="co"># convert from boolean to character</span></span>
<span>        `BCL2 Status` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">`BCL2 Status`</span>, <span class="st">"POS"</span>, <span class="st">"NEG"</span><span class="op">)</span>,</span>
<span>        <span class="co"># adding seq type for compatibility with plotting</span></span>
<span>        seq_type <span class="op">=</span> <span class="st">"genome"</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<p>We also want to ensure the consistent ordering of the Analysis cohort as the annotation track associated with heatmap. By default, the values will be sorted alphabetically, but we will set the ordering more informed biologically by converting that column to a factor:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">metadata</span> <span class="op">&lt;-</span> <span class="va">metadata</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        `Analysis cohort` <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span></span>
<span>            <span class="va">`Analysis cohort`</span>,</span>
<span>            levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>                <span class="st">"denovo-DLBCL"</span>,</span>
<span>                <span class="st">"no-HT"</span>,</span>
<span>                <span class="st">"pre-HT"</span>,</span>
<span>                <span class="st">"post-HT"</span>,</span>
<span>                <span class="st">"COM"</span></span>
<span>            <span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
</div>
<p>We will now also collect mutations for these samples:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_ssm_by_samples.html" class="external-link">get_ssm_by_samples</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    tool_name <span class="op">=</span> <span class="st">"publication"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<p>Now, we need a bed-formatted data frame with coordinates for the regions of interest. We will use the aSHM regions from GAMBLR, and will use the same version as was used in the original paper:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">some_regions</span> <span class="op">&lt;-</span> <span class="va">somatic_hypermutation_locations_GRCh37_v0.2</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span></span>
<span>        <span class="st">"chrom"</span> <span class="op">=</span> <span class="st">"chr_name"</span>,</span>
<span>        <span class="st">"start"</span> <span class="op">=</span> <span class="st">"hg19_start"</span>,</span>
<span>        <span class="st">"end"</span> <span class="op">=</span> <span class="st">"hg19_end"</span>,</span>
<span>        <span class="st">"name"</span> <span class="op">=</span> <span class="st">"gene"</span></span>
<span>    <span class="op">)</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>chrom <span class="op">=</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html" class="external-link">str_remove</a></span><span class="op">(</span><span class="va">chrom</span>, <span class="st">"chr"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#set factor ordering for later</span></span>
<span><span class="va">some_regions</span><span class="op">$</span><span class="va">name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span></span>
<span>    <span class="va">some_regions</span><span class="op">$</span><span class="va">name</span>,</span>
<span>    levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">some_regions</span><span class="op">$</span><span class="va">name</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<p>Now, we can plot the heatmap of mutations using the GAMBLR.viz function <code>prettyMutationDensity</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ashm_heatmap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prettyMutationDensity.html">prettyMutationDensity</a></span><span class="op">(</span></span>
<span>    these_samples_metadata <span class="op">=</span> <span class="va">metadata</span>,</span>
<span>    maf_data <span class="op">=</span> <span class="va">maf</span>,</span>
<span>    cluster_samples <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    regions_bed <span class="op">=</span> <span class="va">some_regions</span>,</span>
<span>    min_bin_recurrence <span class="op">=</span> <span class="fl">10</span>,</span>
<span>    region_fontsize <span class="op">=</span> <span class="fl">5</span>,</span>
<span>    window_size <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>    slide_by <span class="op">=</span> <span class="fl">500</span>,</span>
<span>    orientation <span class="op">=</span> <span class="st">"sample_columns"</span>,</span>
<span>    sortByMetadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Analysis cohort"</span>, <span class="st">"BCL2 Status"</span><span class="op">)</span>,</span>
<span>    metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BCL2 Status"</span>, <span class="st">"Analysis cohort"</span><span class="op">)</span>,</span>
<span>    backgroundColour <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>    customColours <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"Analysis cohort"</span> <span class="op">=</span> <span class="fu">get_gambl_colours</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        <span class="st">"BCL2 Status"</span> <span class="op">=</span> <span class="fu">get_gambl_colours</span><span class="op">(</span><span class="st">"clinical"</span><span class="op">)</span></span>
<span>    <span class="op">)</span>,</span>
<span>    returnEverything <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/draw-dispatch.html" class="external-link">draw</a></span><span class="op">(</span></span>
<span>    <span class="va">ashm_heatmap</span><span class="op">$</span><span class="va">heatmap_object</span>,</span>
<span>    heatmap_legend_side <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>    annotation_legend_side <span class="op">=</span> <span class="st">"bottom"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="cell-output-display">
<div>
<figure><p><img src="recap_figures_files/figure-html/ashm_plot-1.png" width="672"></p>
</figure>
</div>
</div>
</div>
</section><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'light-border',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config);
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script></main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
