<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>ASHM Multi-panel Rainbow Plot — ashm_multi_rainbow_plot • GAMBLR.viz</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="ASHM Multi-panel Rainbow Plot — ashm_multi_rainbow_plot"><meta property="og:description" content="Generates a colourful multi-panel overview of hypermutation in regions of interest across many samples."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.viz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>ASHM Multi-panel Rainbow Plot</h1>
    
    <div class="hidden name"><code>ashm_multi_rainbow_plot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generates a colourful multi-panel overview of hypermutation in regions of interest across many samples.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ashm_multi_rainbow_plot</span><span class="op">(</span></span>
<span>  <span class="va">regions_bed</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  <span class="va">custom_colours</span>,</span>
<span>  classification_column <span class="op">=</span> <span class="st">"lymphgen"</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>regions_bed</dt>
<dd><p>Bed file with chromosome coordinates, should contain columns chr, start, end, name (with these exact names). Not required if selecting from many common regions; bonus regions also exist in grch37.</p></dd>


<dt>these_samples_metadata</dt>
<dd><p>A metadata file already subsetted and arranged on the order you want the samples vertically displayed.</p></dd>


<dt>this_seq_type</dt>
<dd><p>the seqtype you want results back for if `maf_data` is not provided.</p></dd>


<dt>custom_colours</dt>
<dd><p>Provide named vector (or named list of vectors) containing custom annotation colours if you do not want to use standardized pallette.</p></dd>


<dt>classification_column</dt>
<dd><p>Optional. Override default column for assigning the labels used for colouring in the figure.</p></dd>


<dt>maf_data</dt>
<dd><p>An already loaded maf. If not provided, this function will call `get_ssm_by_region`, using the regions supplied into `regions_bed`. Ensure your maf matches the genome projection.</p></dd>


<dt>projection</dt>
<dd><p>Provide genome build; default is grch37. Bonus regions are only available in grch37.</p></dd>


<dt>verbose</dt>
<dd><p>Set to FALSE to prevent printing the full regions bed file to the console. Default is TRUE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Nothing</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The input for this function is a bed-file with the following columns; chr, start, end, name.
Note that for this function to work, the column names must be exactly this.
The user also needs to specify a vector of names (`regions_to_display`) to further control what regions are to be displayed on the returned plot.
It is also possible to exclude specific classifications from the metadata file. This is achieved with `exclude_classifications`.
In addition the user can also use the `metadata` parameter to use an already subset and arranged metadata table.
This function will call [GAMBLR::get_ssm_by_region] if `maf_data` is not called. For more info, refer to the parameter descriptions of this function.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">#get lymphgen colours</span></span></span>
<span class="r-in"><span><span class="va">lymphgen_colours</span> <span class="op">=</span> <span class="fu">GAMBLR.helpers</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.helpers/man/get_gambl_colours.html" class="external-link">get_gambl_colours</a></span><span class="op">(</span><span class="st">"lymphgen"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu">GAMBLR.open</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">==</span><span class="st">"DLBCL"</span>,</span></span>
<span class="r-in"><span>                 <span class="va">seq_type</span><span class="op">==</span><span class="st">"genome"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">lymphgen</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu">check_and_clean_metadata</span><span class="op">(</span><span class="va">metadata</span>,</span></span>
<span class="r-in"><span>                            duplicate_action <span class="op">=</span><span class="st">"keep_first"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">regions_bed</span> <span class="op">=</span> <span class="fu">create_bed_data</span><span class="op">(</span><span class="va">grch37_ashm_regions</span>,</span></span>
<span class="r-in"><span>                              fix_names <span class="op">=</span> <span class="st">"concat"</span>,</span></span>
<span class="r-in"><span>                              concat_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>,<span class="st">"region"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">regions_bed</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">regions_bed</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"BCL6"</span>,<span class="va">name</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">ashm_multi_rainbow_plot</span><span class="op">(</span><span class="va">regions_bed</span>,</span></span>
<span class="r-in"><span>                        <span class="va">metadata</span>,</span></span>
<span class="r-in"><span>                        custom_colours <span class="op">=</span> <span class="va">lymphgen_colours</span>,</span></span>
<span class="r-in"><span>                        verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ashm_multi_rainbow_plot"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "regions_bed provided"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   chrom     start       end             name       region regulatory_comment</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     3 187458526 187464632          BCL6TSS          TSS               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     3 187615533 187625659 BCL6Intergenic-1 Intergenic-1               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     3 187625659 187638101 BCL6Intergenic-2 Intergenic-2               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     3 187657518 187665996 BCL6Intergenic-3 Intergenic-3               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5     3 187675741 187690717 BCL6Intergenic-4 Intergenic-4               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6     3 187690717 187705000 BCL6Intergenic-5 Intergenic-5               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "get_ssm_by_regions"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled SSM calls (.maf) calls in GAMBLR.data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled SSM calls (.maf) calls in GAMBLR.data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running in default mode of any...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       start sample_id      region_name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 187620119  OCI-Ly10 BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 187622610  SU-DHL-4 BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 187615657  FL1003T2 BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 187616207  SP116648 BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 187616296  SP116659 BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 187616515 96-31596T BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 8</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   COO_consensus DHITsig_consensus EBV_status_inf Tumor_Sample_Barcode age_group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> GCB           DHITsig-IND       <span style="color: #BB0000;">NA</span>             00-14595_tumorC      Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> ABC           DHITsigNeg        <span style="color: #BB0000;">NA</span>             00-15201_tumorA      Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> ABC           DHITsigNeg        <span style="color: #BB0000;">NA</span>             00-15201_tumorB      Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> ABC           DHITsigNeg        <span style="color: #BB0000;">NA</span>             00-26427_tumorA      Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> ABC           DHITsigNeg        <span style="color: #BB0000;">NA</span>             00-26427_tumorC      Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> GCB           DHITsigNeg        <span style="color: #BB0000;">NA</span>             01-14774_tumorA      Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 3 more variables: bam_available &lt;lgl&gt;, cohort &lt;chr&gt;, compression &lt;chr&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(sample_id)`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(start)`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(sample_id)`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       start sample_id      region_name COO_consensus DHITsig_consensus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 187620119  OCI-Ly10 BCL6Intergenic-1           ABC        DHITsigNeg</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 187622610  SU-DHL-4 BCL6Intergenic-1           GCB        DHITsigPos</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 187615657  FL1003T2 BCL6Intergenic-1           GCB        DHITsigPos</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 187616207  SP116648 BCL6Intergenic-1          &lt;NA&gt;                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 187616296  SP116659 BCL6Intergenic-1          &lt;NA&gt;                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 187616515 96-31596T BCL6Intergenic-1       UNCLASS        DHITsigNeg</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   EBV_status_inf Tumor_Sample_Barcode age_group bam_available</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   EBV-negative             OCI-Ly10     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   EBV-negative             SU-DHL-4     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           &lt;NA&gt;             FL1003T2     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4           &lt;NA&gt;             SP116648     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5           &lt;NA&gt;             SP116659     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6           &lt;NA&gt;            96-31596T     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               cohort</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   DLBCL_cell_lines</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   DLBCL_cell_lines</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          FL_Kridel</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         DLBCL_ICGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         DLBCL_ICGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 DLBCL_GenomeCanada</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "plotting"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "done"</span>
<span class="r-plt img"><img src="ashm_multi_rainbow_plot-1.png" alt="" width="864" height="864"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#build plot</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="fu">ashm_multi_rainbow_plot</span><span class="op">(</span>regions_to_display <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BCL2-TSS"</span>,</span></span>
<span class="r-in"><span>                                               <span class="st">"MYC-TSS"</span>,</span></span>
<span class="r-in"><span>                                               <span class="st">"SGK1-TSS"</span>,</span></span>
<span class="r-in"><span>                                               <span class="st">"IGL"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                        custom_colours <span class="op">=</span> <span class="va">lymphgen_colours</span>,</span></span>
<span class="r-in"><span>                        this_seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

