<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>ASHM Multi-panel Rainbow Plot — ashm_multi_rainbow_plot • GAMBLR.viz</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ASHM Multi-panel Rainbow Plot — ashm_multi_rainbow_plot"><meta name="description" content="Generates a colourful multi-panel overview of hypermutation in regions of interest across many samples."><meta property="og:description" content="Generates a colourful multi-panel overview of hypermutation in regions of interest across many samples."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-viz.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Installation</h6></li>
    <li><a class="dropdown-item" href="../articles/install.html">Installation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/recap_figures.html">Reproducing figures from GAMBL papers</a></li>
    <li><a class="dropdown-item" href="../articles/forestplot.html">Easy forest plots</a></li>
    <li><a class="dropdown-item" href="../articles/CopyNumberHeatmap.html">Visualizing copy number data as heatmaps</a></li>
    <li><a class="dropdown-item" href="../articles/data_q.html">Exploring data quality</a></li>
    <li><a class="dropdown-item" href="../articles/oncoplot.html">Upgrade your Oncoplots</a></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Metadata and GAMBLR colours</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Help</h6></li>
    <li><a class="dropdown-item" href="../articles/faq.html">Frequently Asked Questions</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>ASHM Multi-panel Rainbow Plot</h1>

      <div class="d-none name"><code>ashm_multi_rainbow_plot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generates a colourful multi-panel overview of hypermutation in regions of interest across many samples.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">ashm_multi_rainbow_plot</span><span class="op">(</span></span>
<span>  <span class="va">regions_bed</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  <span class="va">custom_colours</span>,</span>
<span>  classification_column <span class="op">=</span> <span class="st">"lymphgen"</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-regions-bed">regions_bed<a class="anchor" aria-label="anchor" href="#arg-regions-bed"></a></dt>
<dd><p>Bed file with chromosome coordinates, should contain columns chr, start, end, name (with these exact names). Not required if selecting from many common regions; bonus regions also exist in grch37.</p></dd>


<dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>A metadata file already subsetted and arranged on the order you want the samples vertically displayed.</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>the seqtype you want results back for if `maf_data` is not provided.</p></dd>


<dt id="arg-custom-colours">custom_colours<a class="anchor" aria-label="anchor" href="#arg-custom-colours"></a></dt>
<dd><p>Provide named vector (or named list of vectors) containing custom annotation colours if you do not want to use standardized pallette.</p></dd>


<dt id="arg-classification-column">classification_column<a class="anchor" aria-label="anchor" href="#arg-classification-column"></a></dt>
<dd><p>Optional. Override default column for assigning the labels used for colouring in the figure.</p></dd>


<dt id="arg-maf-data">maf_data<a class="anchor" aria-label="anchor" href="#arg-maf-data"></a></dt>
<dd><p>An already loaded maf. If not provided, this function will call `get_ssm_by_region`, using the regions supplied into `regions_bed`. Ensure your maf matches the genome projection.</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>Provide genome build; default is grch37. Bonus regions are only available in grch37.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Set to FALSE to prevent printing the full regions bed file to the console. Default is TRUE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Nothing</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The input for this function is a bed-file with the following columns; chr, start, end, name.
Note that for this function to work, the column names must be exactly this.
The user also needs to specify a vector of names (`regions_to_display`) to further control what regions are to be displayed on the returned plot.
It is also possible to exclude specific classifications from the metadata file. This is achieved with `exclude_classifications`.
In addition the user can also use the `metadata` parameter to use an already subset and arranged metadata table.
This function will try to obtain mutations internally if `maf_data` is not given.
For more info, refer to the parameter descriptions of this function.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">#get lymphgen colours</span></span></span>
<span class="r-in"><span><span class="va">lymphgen_colours</span> <span class="op">=</span> <span class="fu">GAMBLR.helpers</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.helpers/man/get_gambl_colours.html" class="external-link">get_gambl_colours</a></span><span class="op">(</span><span class="st">"lymphgen"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">metadata</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu">GAMBLR.open</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">==</span><span class="st">"DLBCL"</span>,</span></span>
<span class="r-in"><span>                 <span class="va">seq_type</span><span class="op">==</span><span class="st">"genome"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span>          <span class="fu">check_and_clean_metadata</span><span class="op">(</span><span class="va">.</span>,duplicate_action<span class="op">=</span><span class="st">"keep_first"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>          <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">lymphgen</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Loading required package: dplyr</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Attaching package: ‘dplyr’</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:stats’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     filter, lag</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> The following objects are masked from ‘package:base’:</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>     intersect, setdiff, setequal, union</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Duplicate rows (keeping first occurrence) for 'sample_id' and 'seq_type' have been dropped.</span>
<span class="r-in"><span><span class="va">regions_bed</span> <span class="op">=</span> <span class="fu">GAMBLR.utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.utils/man/create_bed_data.html" class="external-link">create_bed_data</a></span><span class="op">(</span><span class="va">grch37_ashm_regions</span>,</span></span>
<span class="r-in"><span>                              fix_names <span class="op">=</span> <span class="st">"concat"</span>,</span></span>
<span class="r-in"><span>                              concat_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>,<span class="st">"region"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">regions_bed</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">regions_bed</span>,<span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"BCL6"</span>,<span class="va">name</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu">ashm_multi_rainbow_plot</span><span class="op">(</span><span class="va">regions_bed</span>,</span></span>
<span class="r-in"><span>                        <span class="va">metadata</span>,</span></span>
<span class="r-in"><span>                        custom_colours <span class="op">=</span> <span class="va">lymphgen_colours</span>,</span></span>
<span class="r-in"><span>                        verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ashm_multi_rainbow_plot"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "regions_bed provided"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   chrom     start       end             name       region regulatory_comment</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1     3 187458526 187464632          BCL6TSS          TSS               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2     3 187615533 187625659 BCL6Intergenic-1 Intergenic-1               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3     3 187625659 187638101 BCL6Intergenic-2 Intergenic-2               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4     3 187657518 187665996 BCL6Intergenic-3 Intergenic-3               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5     3 187675741 187690717 BCL6Intergenic-4 Intergenic-4               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6     3 187690717 187705000 BCL6Intergenic-5 Intergenic-5               &lt;NA&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "get_ssm_by_regions"</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled SSM calls (.maf) calls in GAMBLR.data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled SSM calls (.maf) calls in GAMBLR.data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Running in default mode of any...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: grch37 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       start sample_id      region_name</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 187620119  OCI-Ly10 BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 187622610  SU-DHL-4 BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 187615657  FL1003T2 BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 187616207  SP116648 BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 187616296  SP116659 BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 187616515 96-31596T BCL6Intergenic-1</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># A tibble: 6 × 8</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   COO_consensus DHITsig_consensus EBV_status_inf Tumor_Sample_Barcode age_group</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">1</span> UNCLASS       DHITsigNeg        <span style="color: #BB0000;">NA</span>             01-14774_tumorB      Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">2</span> ABC           DHITsig-IND       <span style="color: #BB0000;">NA</span>             01-23117_tumorB      Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">3</span> ABC           DHITsigNeg        <span style="color: #BB0000;">NA</span>             04-14093_tumorB      Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">4</span> UNCLASS       DHITsigNeg        <span style="color: #BB0000;">NA</span>             04-24061_tumorB      Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">5</span> UNCLASS       DHITsigPos        <span style="color: #BB0000;">NA</span>             05-22052             Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #BCBCBC;">6</span> UNCLASS       DHITsigPos        <span style="color: #BB0000;">NA</span>             05-22052T            Other    </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> <span style="color: #949494;"># ℹ 3 more variables: bam_available &lt;lgl&gt;, cohort &lt;chr&gt;, compression &lt;chr&gt;</span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(sample_id)`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(start)`</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(sample_id)`</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> genomic_data Object</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Genome Build: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Showing first 10 rows:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       start sample_id      region_name COO_consensus DHITsig_consensus</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1 187620119  OCI-Ly10 BCL6Intergenic-1           ABC        DHITsigNeg</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2 187622610  SU-DHL-4 BCL6Intergenic-1           GCB        DHITsigPos</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3 187615657  FL1003T2 BCL6Intergenic-1           GCB        DHITsigPos</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4 187616207  SP116648 BCL6Intergenic-1          &lt;NA&gt;                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5 187616296  SP116659 BCL6Intergenic-1          &lt;NA&gt;                NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 187616515 96-31596T BCL6Intergenic-1       UNCLASS        DHITsigNeg</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   EBV_status_inf Tumor_Sample_Barcode age_group bam_available</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   EBV-negative             OCI-Ly10     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   EBV-negative             SU-DHL-4     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3           &lt;NA&gt;             FL1003T2     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4           &lt;NA&gt;             SP116648     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5           &lt;NA&gt;             SP116659     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6           &lt;NA&gt;            96-31596T     Other          TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>               cohort</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 1   DLBCL_cell_lines</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 2   DLBCL_cell_lines</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 3          FL_Kridel</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 4         DLBCL_ICGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 5         DLBCL_ICGC</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 6 DLBCL_GenomeCanada</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "plotting"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "done"</span>
<span class="r-plt img"><img src="ashm_multi_rainbow_plot-1.png" alt="" width="864" height="778"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#build plot</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu">ashm_multi_rainbow_plot</span><span class="op">(</span>regions_to_display <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BCL2-TSS"</span>,</span></span>
<span class="r-in"><span>                                               <span class="st">"MYC-TSS"</span>,</span></span>
<span class="r-in"><span>                                               <span class="st">"SGK1-TSS"</span>,</span></span>
<span class="r-in"><span>                                               <span class="st">"IGL"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                        custom_colours <span class="op">=</span> <span class="va">lymphgen_colours</span>,</span></span>
<span class="r-in"><span>                        this_seq_type <span class="op">=</span> <span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

