<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>genome-wide ideogram annotated with SSM and CN information — fancy_ideogram • GAMBLR.viz</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="genome-wide ideogram annotated with SSM and CN information — fancy_ideogram"><meta name="description" content="Generate sample-level ideogram with copy number information, ssm and gene annotations, etc."><meta property="og:description" content="Generate sample-level ideogram with copy number information, ssm and gene annotations, etc."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Introduction</h6></li>
    <li><a class="dropdown-item" href="../articles/why.html">GAMBLR.viz</a></li>
    <li><a class="dropdown-item" href="../articles/install.html"></a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/recap_figures.html">Reproducing figures from GAMBL papers</a></li>
    <li><a class="dropdown-item" href="../articles/forestplot.html">Easy forest plots</a></li>
    <li><a class="dropdown-item" href="../articles/CopyNumberHeatmap.html">Visualizing copy number data as heatmaps</a></li>
    <li><a class="dropdown-item" href="../articles/data_q.html">Exploring data quality</a></li>
    <li><a class="dropdown-item" href="../articles/oncoplot.html">Upgrade your Oncoplots</a></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Metadata and GAMBLR colours</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Help</h6></li>
    <li><a class="dropdown-item" href="../articles/faq.html">Frequently Asked Qestions</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>genome-wide ideogram annotated with SSM and CN information</h1>

      <div class="d-none name"><code>fancy_ideogram.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generate sample-level ideogram with copy number information, ssm and gene annotations, etc.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fancy_ideogram</span><span class="op">(</span></span>
<span>  <span class="va">this_sample_id</span>,</span>
<span>  <span class="va">gene_annotation</span>,</span>
<span>  <span class="va">seg_data</span>,</span>
<span>  seg_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  maf_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  variant_type_col_maf <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  chromosome_col_maf <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  start_col_maf <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  end_col_maf <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  chrom_col_seg <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  start_col_seg <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  end_col_seg <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  cn_col_seg <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">this_sample_id</span><span class="op">)</span>,</span>
<span>  plot_subtitle <span class="op">=</span> <span class="st">"Genome-wide Ideogram (grch37)."</span>,</span>
<span>  <span class="va">intersect_regions</span>,</span>
<span>  include_ssm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  ssm_count <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  coding_only <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-this-sample-id">this_sample_id<a class="anchor" aria-label="anchor" href="#arg-this-sample-id"></a></dt>
<dd><p>Sample to be plotted (for multiple samples, see fancy_multisample_ideogram.</p></dd>


<dt id="arg-gene-annotation">gene_annotation<a class="anchor" aria-label="anchor" href="#arg-gene-annotation"></a></dt>
<dd><p>Annotate ideogram with a set of genes. These genes can either be specified as a vector of characters or a data frame.</p></dd>


<dt id="arg-seg-data">seg_data<a class="anchor" aria-label="anchor" href="#arg-seg-data"></a></dt>
<dd><p>Optional parameter with copy number df already loaded into R.</p></dd>


<dt id="arg-seg-path">seg_path<a class="anchor" aria-label="anchor" href="#arg-seg-path"></a></dt>
<dd><p>Optional parameter with path to external seg like file.</p></dd>


<dt id="arg-maf-data">maf_data<a class="anchor" aria-label="anchor" href="#arg-maf-data"></a></dt>
<dd><p>Optional parameter with maf like df already loaded into R.</p></dd>


<dt id="arg-maf-path">maf_path<a class="anchor" aria-label="anchor" href="#arg-maf-path"></a></dt>
<dd><p>Optional parameter with path to external maf like file.</p></dd>


<dt id="arg-variant-type-col-maf">variant_type_col_maf<a class="anchor" aria-label="anchor" href="#arg-variant-type-col-maf"></a></dt>
<dd><p>Index of column holding Variant Type (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-chromosome-col-maf">chromosome_col_maf<a class="anchor" aria-label="anchor" href="#arg-chromosome-col-maf"></a></dt>
<dd><p>Index of column holding Chromosome (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-start-col-maf">start_col_maf<a class="anchor" aria-label="anchor" href="#arg-start-col-maf"></a></dt>
<dd><p>Index of column with variant start coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-end-col-maf">end_col_maf<a class="anchor" aria-label="anchor" href="#arg-end-col-maf"></a></dt>
<dd><p>Index of column with variant end coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-chrom-col-seg">chrom_col_seg<a class="anchor" aria-label="anchor" href="#arg-chrom-col-seg"></a></dt>
<dd><p>Index of column with chromosome annotations (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-start-col-seg">start_col_seg<a class="anchor" aria-label="anchor" href="#arg-start-col-seg"></a></dt>
<dd><p>Index of column with copy number start coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-end-col-seg">end_col_seg<a class="anchor" aria-label="anchor" href="#arg-end-col-seg"></a></dt>
<dd><p>Index of column with copy number end coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-cn-col-seg">cn_col_seg<a class="anchor" aria-label="anchor" href="#arg-cn-col-seg"></a></dt>
<dd><p>Index of column holding copy number information (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-plot-title">plot_title<a class="anchor" aria-label="anchor" href="#arg-plot-title"></a></dt>
<dd><p>Title of plot (default to sample ID).</p></dd>


<dt id="arg-plot-subtitle">plot_subtitle<a class="anchor" aria-label="anchor" href="#arg-plot-subtitle"></a></dt>
<dd><p>Optional argument for plot subtitle.</p></dd>


<dt id="arg-intersect-regions">intersect_regions<a class="anchor" aria-label="anchor" href="#arg-intersect-regions"></a></dt>
<dd><p>Optional parameter for subset variant calls to specific regions. Should be either a vector of characters (chr:start-end) or data frame with regions.</p></dd>


<dt id="arg-include-ssm">include_ssm<a class="anchor" aria-label="anchor" href="#arg-include-ssm"></a></dt>
<dd><p>Set to TRUE to plot SSMs (dels and ins).</p></dd>


<dt id="arg-ssm-count">ssm_count<a class="anchor" aria-label="anchor" href="#arg-ssm-count"></a></dt>
<dd><p>Optional parameter to summarize n variants per chromosome, inlcude_ssm must be set to TRUE.</p></dd>


<dt id="arg-coding-only">coding_only<a class="anchor" aria-label="anchor" href="#arg-coding-only"></a></dt>
<dd><p>Optional. Set to TRUE to restrict to plotting only coding mutations.</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>Seq type for returned CN segments. One of "genome" (default) or "capture".</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A plot as a ggplot object (grob).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function generates genome-wide ideograms, visualizing SSM data as well as CN segments.
It is also possible to superimpose the plot with gene annotations. Offering a comprehensive overview of all SSM and CN segments of different aneuploidy.
The plotting of SSM can be toggled with setting `include_ssm` to TRUE. If so, it is also possible to count the number of SSMs per chromosome with `ssm_count = TRUE`.
To get data for plotting, there are a few different options available; like all `fanncy_x_plots` a sample ID can be provided to the `this_sample_id`
parameter. If done so, the function will retrieve data (SSm and CN segments) by wrapping the appropriate functions.
This data can also be provided with `seg_data`, `seg_path`, `maf_data` and `maf_path`.
For more info on how to run with these parameters, refer to the parameter descriptions.
In order to annotate the ideogram with genes, simply give the `gene_annotations` parameter a set of genes as a vector of characters or a data frame with gene names in the first column.
Another useful parameter for restricting the plotted regions is to call the function with `intersect_regions`.
This parameter takes a vector of characters or a data frame with regions that the plotted calls are restricted to.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co">#build plot</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu">fancy_ideogram</span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"DOHH-2"</span>,</span></span>
<span class="r-in"><span>               gene_annotation <span class="op">=</span> <span class="st">"MYC"</span>,</span></span>
<span class="r-in"><span>               plot_title <span class="op">=</span> <span class="st">"Sample-level Ideogram Example"</span>,</span></span>
<span class="r-in"><span>               plot_subtitle <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

