<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Variant size distribution plot — fancy_v_sizedis • GAMBLR.viz</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Variant size distribution plot — fancy_v_sizedis"><meta name="description" content="Generate a violin plot showing variant (SSM or SVs) size distributions for selected contigs."><meta property="og:description" content="Generate a violin plot showing variant (SSM or SVs) size distributions for selected contigs."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-viz.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Installation</h6></li>
    <li><a class="dropdown-item" href="../articles/install.html">Installation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/recap_figures.html">Reproducing figures from GAMBL papers</a></li>
    <li><a class="dropdown-item" href="../articles/forestplot.html">Easy forest plots</a></li>
    <li><a class="dropdown-item" href="../articles/CopyNumberHeatmap.html">Visualizing copy number data as heatmaps</a></li>
    <li><a class="dropdown-item" href="../articles/data_q.html">Exploring data quality</a></li>
    <li><a class="dropdown-item" href="../articles/oncoplot.html">Upgrade your Oncoplots</a></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Metadata and GAMBLR colours</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Help</h6></li>
    <li><a class="dropdown-item" href="../articles/faq.html">Frequently Asked Questions</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Variant size distribution plot</h1>

      <div class="d-none name"><code>fancy_v_sizedis.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Generate a violin plot showing variant (SSM or SVs) size distributions for selected contigs.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fancy_v_sizedis</span><span class="op">(</span></span>
<span>  <span class="va">this_sample_id</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  maf_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ssm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  min_vaf <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  variant_type_col <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  chromosome_col <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  start_col <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  end_col <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">this_sample_id</span><span class="op">)</span>,</span>
<span>  plot_subtitle <span class="op">=</span> <span class="st">"Variant Size Distribution"</span>,</span>
<span>  scale_value <span class="op">=</span> <span class="st">"width"</span>,</span>
<span>  log_10 <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_trim <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-this-sample-id">this_sample_id<a class="anchor" aria-label="anchor" href="#arg-this-sample-id"></a></dt>
<dd><p>Sample to be plotted.</p></dd>


<dt id="arg-maf-data">maf_data<a class="anchor" aria-label="anchor" href="#arg-maf-data"></a></dt>
<dd><p>Optional parameter with maf like df already loaded into R.</p></dd>


<dt id="arg-maf-path">maf_path<a class="anchor" aria-label="anchor" href="#arg-maf-path"></a></dt>
<dd><p>Optional parameter with path to external maf like file.</p></dd>


<dt id="arg-ssm">ssm<a class="anchor" aria-label="anchor" href="#arg-ssm"></a></dt>
<dd><p>Set to FALSE to get plotting data from get_combined_sv (SVs). Default value is TRUE (plots SSM retrieved from annotate_cn_by_ssm$maf).</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>Genome build for returned variants (only applicable for ssm = FALSE).</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>Seq type for returned CN segments. One of "genome" (default) or "capture".</p></dd>


<dt id="arg-min-vaf">min_vaf<a class="anchor" aria-label="anchor" href="#arg-min-vaf"></a></dt>
<dd><p>The minimum tumour VAF for a SV to be returned. Recommended: 0 (only applicable for ssm = FALSE).</p></dd>


<dt id="arg-variant-type-col">variant_type_col<a class="anchor" aria-label="anchor" href="#arg-variant-type-col"></a></dt>
<dd><p>Index of column holding Variant Type (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-chromosome-col">chromosome_col<a class="anchor" aria-label="anchor" href="#arg-chromosome-col"></a></dt>
<dd><p>Index of column holding Chromosome (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-start-col">start_col<a class="anchor" aria-label="anchor" href="#arg-start-col"></a></dt>
<dd><p>Index of column with variant start coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-end-col">end_col<a class="anchor" aria-label="anchor" href="#arg-end-col"></a></dt>
<dd><p>Index of column with variant end coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-plot-title">plot_title<a class="anchor" aria-label="anchor" href="#arg-plot-title"></a></dt>
<dd><p>Title of plot (default to sample ID).</p></dd>


<dt id="arg-plot-subtitle">plot_subtitle<a class="anchor" aria-label="anchor" href="#arg-plot-subtitle"></a></dt>
<dd><p>Subtitle for created plot.</p></dd>


<dt id="arg-scale-value">scale_value<a class="anchor" aria-label="anchor" href="#arg-scale-value"></a></dt>
<dd><p>Scale type for violin plot, accepted values are "area", "width", and "count", default is "count.</p></dd>


<dt id="arg-log-">log_10<a class="anchor" aria-label="anchor" href="#arg-log-"></a></dt>
<dd><p>Boolean statement for y-axis, default is TRUE.</p></dd>


<dt id="arg-plot-trim">plot_trim<a class="anchor" aria-label="anchor" href="#arg-plot-trim"></a></dt>
<dd><p>If TRUE, trim the tails of the violins to the range of the data. If FALSE (default), don't trim the tails.</p></dd>


<dt id="arg-chr-select">chr_select<a class="anchor" aria-label="anchor" href="#arg-chr-select"></a></dt>
<dd><p>vector of chromosomes to be included in plot, defaults to autosomes.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A plot as a ggplot object (grob).</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>Function for plotting variant size distributions. This function takes either a sample ID given to the `this_sample` parameter.
In addition, the function can also accept an already loaded MAF or MAF-like object given to the `maf_data` parameter.
As a third option, the function can also read a maf from disk (provide path to maf with `maf_path`).
A collection of convenient filtering and data subsetting parameters are also available for this function.
This plot can also deal with SVs as well as SSM data. To control this, please use the `ssm` parameter. If set to TRUE and if `this_sample` is called,
the function gets data with get_ssm_by_sample and if set to FALSE, the function calls `get_manta_sv` to get SV calls for plotting.
If the user calls either `maf_data` or `maf_path`, there are a collection of parameters available for specifying
the relevant columns in the given data frame (`variant_type_col`, `chhromosome_col`, `start_col`, `end_col`). It is also possible to
restrict the returned plot to any given chromosome. This is done with the `chr_select` parameter (default is all autosomes).
In addition, plot aesthetics can also be controlled with `plot_title`, `plot_subtitle`, `scale_value`, `log10`, and `trim`.
For more info on how to run with these parameters, refer to the parameter descriptions.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.data</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#plot SSM size distributions:</span></span></span>
<span class="r-in"><span><span class="fu">fancy_v_sizedis</span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"DOHH-2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

