<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Variant size distribution plot — fancy_v_sizedis • GAMBLR.viz</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Variant size distribution plot — fancy_v_sizedis"><meta property="og:description" content="Generate a violin plot showing variant (SSM or SVs) size distributions for selected contigs."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.viz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Variant size distribution plot</h1>

    <div class="hidden name"><code>fancy_v_sizedis.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Generate a violin plot showing variant (SSM or SVs) size distributions for selected contigs.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fancy_v_sizedis</span><span class="op">(</span></span>
<span>  <span class="va">this_sample_id</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  maf_path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  ssm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  projection <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="st">"genome"</span>,</span>
<span>  min_vaf <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  variant_type_col <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  chromosome_col <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  start_col <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  end_col <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">this_sample_id</span><span class="op">)</span>,</span>
<span>  plot_subtitle <span class="op">=</span> <span class="st">"Variant Size Distribution"</span>,</span>
<span>  scale_value <span class="op">=</span> <span class="st">"width"</span>,</span>
<span>  log_10 <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  plot_trim <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  chr_select <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-this-sample-id">this_sample_id<a class="anchor" aria-label="anchor" href="#arg-this-sample-id"></a></dt>
<dd><p>Sample to be plotted.</p></dd>


<dt id="arg-maf-data">maf_data<a class="anchor" aria-label="anchor" href="#arg-maf-data"></a></dt>
<dd><p>Optional parameter with maf like df already loaded into R.</p></dd>


<dt id="arg-maf-path">maf_path<a class="anchor" aria-label="anchor" href="#arg-maf-path"></a></dt>
<dd><p>Optional parameter with path to external maf like file.</p></dd>


<dt id="arg-ssm">ssm<a class="anchor" aria-label="anchor" href="#arg-ssm"></a></dt>
<dd><p>Set to FALSE to get plotting data from get_combined_sv (SVs). Default value is TRUE (plots SSM retrieved from annotate_cn_by_ssm$maf).</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>Genome build for returned variants (only applicable for ssm = FALSE).</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>Seq type for returned CN segments. One of "genome" (default) or "capture".</p></dd>


<dt id="arg-min-vaf">min_vaf<a class="anchor" aria-label="anchor" href="#arg-min-vaf"></a></dt>
<dd><p>The minimum tumour VAF for a SV to be returned. Recommended: 0 (only applicable for ssm = FALSE).</p></dd>


<dt id="arg-variant-type-col">variant_type_col<a class="anchor" aria-label="anchor" href="#arg-variant-type-col"></a></dt>
<dd><p>Index of column holding Variant Type (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-chromosome-col">chromosome_col<a class="anchor" aria-label="anchor" href="#arg-chromosome-col"></a></dt>
<dd><p>Index of column holding Chromosome (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-start-col">start_col<a class="anchor" aria-label="anchor" href="#arg-start-col"></a></dt>
<dd><p>Index of column with variant start coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-end-col">end_col<a class="anchor" aria-label="anchor" href="#arg-end-col"></a></dt>
<dd><p>Index of column with variant end coordinates (to be used with either maf_data or maf_path).</p></dd>


<dt id="arg-plot-title">plot_title<a class="anchor" aria-label="anchor" href="#arg-plot-title"></a></dt>
<dd><p>Title of plot (default to sample ID).</p></dd>


<dt id="arg-plot-subtitle">plot_subtitle<a class="anchor" aria-label="anchor" href="#arg-plot-subtitle"></a></dt>
<dd><p>Subtitle for created plot.</p></dd>


<dt id="arg-scale-value">scale_value<a class="anchor" aria-label="anchor" href="#arg-scale-value"></a></dt>
<dd><p>Scale type for violin plot, accepted values are "area", "width", and "count", default is "count.</p></dd>


<dt id="arg-log-">log_10<a class="anchor" aria-label="anchor" href="#arg-log-"></a></dt>
<dd><p>Boolean statement for y-axis, default is TRUE.</p></dd>


<dt id="arg-plot-trim">plot_trim<a class="anchor" aria-label="anchor" href="#arg-plot-trim"></a></dt>
<dd><p>If TRUE, trim the tails of the violins to the range of the data. If FALSE (default), don't trim the tails.</p></dd>


<dt id="arg-chr-select">chr_select<a class="anchor" aria-label="anchor" href="#arg-chr-select"></a></dt>
<dd><p>vector of chromosomes to be included in plot, defaults to autosomes.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A plot as a ggplot object (grob).</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Function for plotting variant size distributions. This function takes either a sample ID given to the `this_sample` parameter.
In addition, the function can also accept an already loaded MAF or MAF-like object given to the `maf_data` parameter.
As a third option, the function can also read a maf from disk (provide path to maf with `maf_path`).
A collection of convenient filtering and data subsetting parameters are also available for this function.
This plot can also deal with SVs as well as SSM data. To control this, please use the `ssm` parameter. If set to TRUE and if `this_sample` is called,
the function gets data with get_ssm_by_sample and if set to FALSE, the function calls `get_manta_sv` to get SV calls for plotting.
If the user calls either `maf_data` or `maf_path`, there are a collection of parameters available for specifying
the relevant columns in the given data frame (`variant_type_col`, `chhromosome_col`, `start_col`, `end_col`). It is also possible to
restrict the returned plot to any given chromosome. This is done with the `chr_select` parameter (default is all autosomes).
In addition, plot aesthetics can also be controlled with `plot_title`, `plot_subtitle`, `scale_value`, `log10`, and `trim`.
For more info on how to run with these parameters, refer to the parameter descriptions.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.data</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#plot SSM size distributions:</span></span></span>
<span class="r-in"><span><span class="fu">fancy_v_sizedis</span><span class="op">(</span>this_sample_id <span class="op">=</span> <span class="st">"DOHH-2"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

