<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Pretty Copy Number Heatmap — pretty_CN_heatmap • GAMBLR.viz</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Pretty Copy Number Heatmap — pretty_CN_heatmap"><meta property="og:description" content="Pretty Copy Number Heatmap"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.viz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Pretty Copy Number Heatmap</h1>

    <div class="hidden name"><code>pretty_CN_heatmap.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Pretty Copy Number Heatmap</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">pretty_CN_heatmap</span><span class="op">(</span></span>
<span>  <span class="va">cn_state_matrix</span>,</span>
<span>  scale_by_sample <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">these_samples_metadata</span>,</span>
<span>  keep_sample_order <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span><span class="op">)</span>,</span>
<span>  <span class="va">expressionColumns</span>,</span>
<span>  genome_build <span class="op">=</span> <span class="st">"grch37"</span>,</span>
<span>  cluster_columns <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_row_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">keep_these_chromosomes</span>,</span>
<span>  <span class="va">hide_these_chromosomes</span>,</span>
<span>  <span class="va">keep_these_bins</span>,</span>
<span>  <span class="va">hide_annotations</span>,</span>
<span>  <span class="va">sortByBins</span>,</span>
<span>  <span class="va">sortByGenes</span>,</span>
<span>  <span class="va">splitByBinState</span>,</span>
<span>  sortByPGA <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">sortByMetadataColumns</span>,</span>
<span>  <span class="va">labelTheseGenes</span>,</span>
<span>  <span class="va">labelTheseCytobands</span>,</span>
<span>  <span class="va">highlightTheseRegions</span>,</span>
<span>  bin_label_fontsize <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  bin_label_nudge <span class="op">=</span> <span class="fl">1.03</span>,</span>
<span>  bin_label_rotation <span class="op">=</span> <span class="fl">45</span>,</span>
<span>  drop_if_PGA_below <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  drop_if_PGA_above <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  <span class="va">focus_on_these_bins</span>,</span>
<span>  <span class="va">geneBoxPlot</span>,</span>
<span>  show_bottom_annotation_name <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  bottom_annotation_name_side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>  left_annotation_name_side <span class="op">=</span> <span class="st">"top"</span>,</span>
<span>  <span class="va">bin_labels</span>,</span>
<span>  legend_direction <span class="op">=</span> <span class="st">"horizontal"</span>,</span>
<span>  legend_position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>  legend_row <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  legend_col <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  metadataBarFontsize <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  metadataBarHeight <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  boxplot_orientation <span class="op">=</span> <span class="st">"vertical"</span>,</span>
<span>  return_data <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  drop_bin_if_sd_below <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  flip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  max_CN_allowed <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  rotate <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  width <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  height <span class="op">=</span> <span class="fl">6</span>,</span>
<span>  <span class="va">cluster_samples</span>,</span>
<span>  <span class="va">cluster_regions</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-cn-state-matrix">cn_state_matrix<a class="anchor" aria-label="anchor" href="#arg-cn-state-matrix"></a></dt>
<dd><p>The output of get_cn_states</p></dd>


<dt id="arg-scale-by-sample">scale_by_sample<a class="anchor" aria-label="anchor" href="#arg-scale-by-sample"></a></dt>
<dd><p>Set to TRUE to scale CN values within each sample_id</p></dd>


<dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>The output of get_gambl_metadata</p></dd>


<dt id="arg-keep-sample-order">keep_sample_order<a class="anchor" aria-label="anchor" href="#arg-keep-sample-order"></a></dt>
<dd><p>FALSE. Set to TRUE to ensure samples are in the same order as in the metadata</p></dd>


<dt id="arg-metadatacolumns">metadataColumns<a class="anchor" aria-label="anchor" href="#arg-metadatacolumns"></a></dt>
<dd><p>One or more columns from the metadata you want to display beside the heatmap</p></dd>


<dt id="arg-expressioncolumns">expressionColumns<a class="anchor" aria-label="anchor" href="#arg-expressioncolumns"></a></dt>
<dd><p>Optional: One or more columns from the metadata that include gene expression values you want shown</p></dd>


<dt id="arg-cluster-columns">cluster_columns<a class="anchor" aria-label="anchor" href="#arg-cluster-columns"></a></dt>
<dd><p>Set to TRUE to enable clustering of genomic regions (columns) based on their CN value across all patients in the heatmap</p></dd>


<dt id="arg-cluster-rows">cluster_rows<a class="anchor" aria-label="anchor" href="#arg-cluster-rows"></a></dt>
<dd><p>Set to TRUE to enable clustering of genomic regions (columns) based on their CN value across all regions in the heatmap</p></dd>


<dt id="arg-show-row-names">show_row_names<a class="anchor" aria-label="anchor" href="#arg-show-row-names"></a></dt>
<dd><p>Set to TRUE to display the ID of every bin (region) shown in the heatmap</p></dd>


<dt id="arg-show-column-names">show_column_names<a class="anchor" aria-label="anchor" href="#arg-show-column-names"></a></dt>
<dd><p>Set to TRUE to display the sample_id of every sample shown in the heatmap</p></dd>


<dt id="arg-keep-these-chromosomes">keep_these_chromosomes<a class="anchor" aria-label="anchor" href="#arg-keep-these-chromosomes"></a></dt>
<dd><p>A vector of chromosome names to include (all others will be excluded)</p></dd>


<dt id="arg-hide-these-chromosomes">hide_these_chromosomes<a class="anchor" aria-label="anchor" href="#arg-hide-these-chromosomes"></a></dt>
<dd><p>A vector of chromosome names to exclude (all others will be included unless keep_these_chromosomes is specified)</p></dd>


<dt id="arg-keep-these-bins">keep_these_bins<a class="anchor" aria-label="anchor" href="#arg-keep-these-bins"></a></dt>
<dd><p>A vector of bin names to include (all others will be excluded)</p></dd>


<dt id="arg-hide-annotations">hide_annotations<a class="anchor" aria-label="anchor" href="#arg-hide-annotations"></a></dt>
<dd><p>A vector of annotation names to suppress from legends in the plot</p></dd>


<dt id="arg-sortbybins">sortByBins<a class="anchor" aria-label="anchor" href="#arg-sortbybins"></a></dt>
<dd><p>Optional: A vector containing one or more names of genomic bins that will be used to order the heatmap rows.</p></dd>


<dt id="arg-splitbybinstate">splitByBinState<a class="anchor" aria-label="anchor" href="#arg-splitbybinstate"></a></dt>
<dd><p>Optional: A single genomic bin that will be used to split the heatmap based on the CN state of that bin</p></dd>


<dt id="arg-sortbypga">sortByPGA<a class="anchor" aria-label="anchor" href="#arg-sortbypga"></a></dt>
<dd><p>Optional: Sort the rows based on percent genome altered (PGA) instead of the other options</p></dd>


<dt id="arg-sortbymetadatacolumns">sortByMetadataColumns<a class="anchor" aria-label="anchor" href="#arg-sortbymetadatacolumns"></a></dt>
<dd><p>A vector containing one or more names of columns from your metadata that will be used to order the rows overall or within slices (if combined with splitByBinState or geneBoxPlot)</p></dd>


<dt id="arg-labelthesegenes">labelTheseGenes<a class="anchor" aria-label="anchor" href="#arg-labelthesegenes"></a></dt>
<dd><p>A vector of Hugo gene symbols whose location will be indicated on the top of the heatmap</p></dd>


<dt id="arg-bin-label-fontsize">bin_label_fontsize<a class="anchor" aria-label="anchor" href="#arg-bin-label-fontsize"></a></dt>
<dd><p>Font size for the gene labels (default 5)</p></dd>


<dt id="arg-bin-label-nudge">bin_label_nudge<a class="anchor" aria-label="anchor" href="#arg-bin-label-nudge"></a></dt>
<dd><p>Increase or decrease this value to shift the gene labels up/down (default 1.03)</p></dd>


<dt id="arg-bin-label-rotation">bin_label_rotation<a class="anchor" aria-label="anchor" href="#arg-bin-label-rotation"></a></dt>
<dd><p>Rotate the direction of the bin label. Default is 45.</p></dd>


<dt id="arg-drop-if-pga-below">drop_if_PGA_below<a class="anchor" aria-label="anchor" href="#arg-drop-if-pga-below"></a></dt>
<dd><p>Lower limit for proportion of genome altered (PGA). Samples below this value will be dropped (default 0)</p></dd>


<dt id="arg-drop-if-pga-above">drop_if_PGA_above<a class="anchor" aria-label="anchor" href="#arg-drop-if-pga-above"></a></dt>
<dd><p>Upper limit for proportion of genome altered (PGA). Samples above this value will be dropped (default 1)</p></dd>


<dt id="arg-focus-on-these-bins">focus_on_these_bins<a class="anchor" aria-label="anchor" href="#arg-focus-on-these-bins"></a></dt>
<dd><p>Mask all regions outside these bins (set CN to 0). Useful for visualizing GISTIC results.</p></dd>


<dt id="arg-geneboxplot">geneBoxPlot<a class="anchor" aria-label="anchor" href="#arg-geneboxplot"></a></dt>
<dd><p>Optional: Specify the Hugo symbol of a single gene to embed box plots adjacent to the heatmap. Expression data for this gene must be present in the metadata in a column of the same name.</p></dd>


<dt id="arg-show-bottom-annotation-name">show_bottom_annotation_name<a class="anchor" aria-label="anchor" href="#arg-show-bottom-annotation-name"></a></dt>
<dd><p>set to TRUE to label the bottom annotation tracks with more details</p></dd>


<dt id="arg-bottom-annotation-name-side">bottom_annotation_name_side<a class="anchor" aria-label="anchor" href="#arg-bottom-annotation-name-side"></a></dt>
<dd><p>If using show_bottom_annotation_name, set this to "left" or "right" to relocate the names</p></dd>


<dt id="arg-left-annotation-name-side">left_annotation_name_side<a class="anchor" aria-label="anchor" href="#arg-left-annotation-name-side"></a></dt>
<dd><p>Which side to put the name of the metadata annotations (top or bottom)</p></dd>


<dt id="arg-bin-labels">bin_labels<a class="anchor" aria-label="anchor" href="#arg-bin-labels"></a></dt>
<dd><p>Instead of automatically labeling genes, you can instead explicitly provide a list of labels for any bins in the heatmap. The names of each element should match a bin. The value of each element is the label that will be shown. This option can be used to skip gene location look-ups (see examples).</p></dd>


<dt id="arg-legend-direction">legend_direction<a class="anchor" aria-label="anchor" href="#arg-legend-direction"></a></dt>
<dd><p>Which orientation to use for the legend</p></dd>


<dt id="arg-legend-position">legend_position<a class="anchor" aria-label="anchor" href="#arg-legend-position"></a></dt>
<dd><p>Where to put the legend</p></dd>


<dt id="arg-legend-row">legend_row<a class="anchor" aria-label="anchor" href="#arg-legend-row"></a></dt>
<dd><p>How many rows for the legend layout</p></dd>


<dt id="arg-legend-col">legend_col<a class="anchor" aria-label="anchor" href="#arg-legend-col"></a></dt>
<dd><p>How many columns for the legend layout</p></dd>


<dt id="arg-boxplot-orientation">boxplot_orientation<a class="anchor" aria-label="anchor" href="#arg-boxplot-orientation"></a></dt>
<dd><p>Either "horizontal" or "vertical" (default: horizontal)</p></dd>


<dt id="arg-return-data">return_data<a class="anchor" aria-label="anchor" href="#arg-return-data"></a></dt>
<dd><p>Specify TRUE to get some of the internal data back including the heatmap object</p></dd>


<dt id="arg-drop-bin-if-sd-below">drop_bin_if_sd_below<a class="anchor" aria-label="anchor" href="#arg-drop-bin-if-sd-below"></a></dt>
<dd><p>Force bins with standard deviation below this value to be excluded</p></dd>


<dt id="arg-flip">flip<a class="anchor" aria-label="anchor" href="#arg-flip"></a></dt>
<dd><p>Optionally, flip the rows/columns of resulting heatmap. Default is FALSE.</p></dd>


<dt id="arg-verbose">verbose<a class="anchor" aria-label="anchor" href="#arg-verbose"></a></dt>
<dd><p>Control verbosity of the console output. Default is FALSE.</p></dd>


<dt id="arg-width">width<a class="anchor" aria-label="anchor" href="#arg-width"></a></dt>
<dd><p>Set the width of the heatmap. Default is 10.</p></dd>


<dt id="arg-cluster-samples">cluster_samples<a class="anchor" aria-label="anchor" href="#arg-cluster-samples"></a></dt>
<dd><p>More intuitive alias for cluster_rows, especially when combining with rotate = TRUE</p></dd>


<dt id="arg-cluster-regions">cluster_regions<a class="anchor" aria-label="anchor" href="#arg-cluster-regions"></a></dt>
<dd><p>More intuitive alias for cluster_columns</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>list (when return_data = TRUE)</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co">#get some metadata for subsetting the data to just one pathology (DLBCL)</span></span></span>
<span class="r-in"><span><span class="va">dlbcl_genome_meta</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>                    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span><span class="op">==</span><span class="st">"DLBCL"</span>,</span></span>
<span class="r-in"><span>                    <span class="va">seq_type</span><span class="op">==</span><span class="st">"genome"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#remove any duplicate sample_id/seq_type combinations</span></span></span>
<span class="r-in"><span><span class="va">meta_clean</span> <span class="op">=</span> <span class="fu">check_and_clean_metadata</span><span class="op">(</span><span class="va">dlbcl_genome_meta</span>,</span></span>
<span class="r-in"><span>                                      duplicate_action <span class="op">=</span> <span class="st">"keep_first"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create the copy number matrix using the helper functions</span></span></span>
<span class="r-in"><span><span class="va">all_segments</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_cn_segments.html" class="external-link">get_cn_segments</a></span><span class="op">(</span>these <span class="op">=</span> <span class="va">meta_clean</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> dummy segments are not annotated in the inputs</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> fill_missing_with parameter will be ignored</span>
<span class="r-in"><span><span class="va">dlbcl_cn_binned</span> <span class="op">=</span> <span class="fu">segmented_data_to_cn_matrix</span><span class="op">(</span></span></span>
<span class="r-in"><span>                                  seg_data <span class="op">=</span> <span class="va">all_segments</span>,</span></span>
<span class="r-in"><span>                                  strategy<span class="op">=</span><span class="st">"auto_split"</span>,</span></span>
<span class="r-in"><span>                                  n_bins_split<span class="op">=</span><span class="fl">1300</span>,</span></span>
<span class="r-in"><span>                                  these_samples_metadata <span class="op">=</span> <span class="va">meta_clean</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Generate a basic genome-wide CN heatmap</span></span></span>
<span class="r-in"><span><span class="fu">pretty_CN_heatmap</span><span class="op">(</span>cn_state_matrix<span class="op">=</span><span class="va">dlbcl_cn_binned</span>,</span></span>
<span class="r-in"><span>                  these_samples_metadata <span class="op">=</span> <span class="va">meta_clean</span>,</span></span>
<span class="r-in"><span>                  hide_annotations <span class="op">=</span> <span class="st">"chromosome"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The input is a data frame-like object, convert it to a matrix.</span>
<span class="r-plt img"><img src="pretty_CN_heatmap-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Disable row (sample) clustering and restrict to a few chromosomes</span></span></span>
<span class="r-in"><span><span class="co"># and highlight some genes of interest</span></span></span>
<span class="r-in"><span><span class="fu">pretty_CN_heatmap</span><span class="op">(</span>cn_state_matrix<span class="op">=</span><span class="va">dlbcl_cn_binned</span>,</span></span>
<span class="r-in"><span>  these_samples_metadata <span class="op">=</span> <span class="va">meta_clean</span>,</span></span>
<span class="r-in"><span>  hide_annotations <span class="op">=</span> <span class="st">"chromosomes"</span>,</span></span>
<span class="r-in"><span>  keep_these_chromosomes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"9"</span>,<span class="st">"17"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  cluster_rows<span class="op">=</span><span class="cn">F</span>,</span></span>
<span class="r-in"><span>  labelTheseGenes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CDKN2A"</span>,<span class="st">"TP53"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> mapping genes to bins</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>The input is a data frame-like object, convert it to a matrix.</span>
<span class="r-plt img"><img src="pretty_CN_heatmap-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span> <span class="co"># get gene expression data</span></span></span>
<span class="r-in"><span> <span class="va">gene_exp_all</span> <span class="op">=</span> <span class="fu">get_gene_expression</span><span class="op">(</span>all_genes<span class="op">=</span><span class="cn">T</span>,</span></span>
<span class="r-in"><span>                                    lazy_join<span class="op">=</span><span class="cn">T</span>,</span></span>
<span class="r-in"><span>                                     arbitrarily_pick <span class="op">=</span> <span class="cn">T</span>,</span></span>
<span class="r-in"><span>                                     HGNC<span class="op">=</span><span class="cn">T</span>,format<span class="op">=</span><span class="st">"wide"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span> <span class="va">genome_meta_exp</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">seq_type</span><span class="op">==</span><span class="st">"genome"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">sample_id</span>,<span class="va">pathology</span>,<span class="va">lymphgen</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">gene_exp_all</span>,<span class="op">-</span><span class="va">sample_id</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    by<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sample_id"</span><span class="op">=</span><span class="st">"genome_sample_id"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>     <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">MYC</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span><span class="co"># Include gene expression data and embed a box plot showing the expression of one gene across different CN states</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span> <span class="fu">pretty_CN_heatmap</span><span class="op">(</span>cn_state_matrix<span class="op">=</span><span class="va">all_states_binned</span>,</span></span>
<span class="r-in"><span>   these_samples_metadata <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">genome_meta_exp</span>,<span class="va">pathology</span><span class="op">==</span><span class="st">"DLBCL"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>   hide_annotations <span class="op">=</span> <span class="st">"chromosomes"</span>,</span></span>
<span class="r-in"><span>   cluster_rows<span class="op">=</span><span class="cn">F</span>,</span></span>
<span class="r-in"><span>   geneBoxPlot <span class="op">=</span> <span class="st">"TP53"</span>,</span></span>
<span class="r-in"><span>   boxplot_orientation<span class="op">=</span><span class="st">"horizontal"</span>,bin_label_fontsize <span class="op">=</span> <span class="fl">9</span>,bin_label_nudge <span class="op">=</span> <span class="fl">19</span></span></span>
<span class="r-in"><span> <span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

