<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Chromosome Plot — prettyChromoplot • GAMBLR.viz</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Chromosome Plot — prettyChromoplot"><meta name="description" content="Use GISTIC2.0 scores output to reproduce maftools::chromoplot with more flexibility."><meta property="og:description" content="Use GISTIC2.0 scores output to reproduce maftools::chromoplot with more flexibility."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/GAMBLR-viz.html">Get started</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Installation</h6></li>
    <li><a class="dropdown-item" href="../articles/install.html">Installation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Guides</h6></li>
    <li><a class="dropdown-item" href="../articles/recap_figures.html">Reproducing figures from GAMBL papers</a></li>
    <li><a class="dropdown-item" href="../articles/forestplot.html">Easy forest plots</a></li>
    <li><a class="dropdown-item" href="../articles/CopyNumberHeatmap.html">Visualizing copy number data as heatmaps</a></li>
    <li><a class="dropdown-item" href="../articles/data_q.html">Exploring data quality</a></li>
    <li><a class="dropdown-item" href="../articles/oncoplot.html">Upgrade your Oncoplots</a></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Metadata and GAMBLR colours</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Help</h6></li>
    <li><a class="dropdown-item" href="../articles/faq.html">Frequently Asked Questions</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Chromosome Plot</h1>

      <div class="d-none name"><code>prettyChromoplot.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Use GISTIC2.0 scores output to reproduce maftools::chromoplot with more flexibility.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">prettyChromoplot</span><span class="op">(</span></span>
<span>  <span class="va">scores_path</span>,</span>
<span>  <span class="va">scores_df</span>,</span>
<span>  <span class="va">labels_bed</span>,</span>
<span>  default_gene_set <span class="op">=</span> <span class="st">"oncogenes"</span>,</span>
<span>  <span class="va">genome_build</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  adjust_amps <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  adjust_dels <span class="op">=</span> <span class="fl">2.75</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  force_pull <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  segment.curvature <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  segment.ncp <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  segment.angle <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  hide_neutral <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-scores-path">scores_path<a class="anchor" aria-label="anchor" href="#arg-scores-path"></a></dt>
<dd><p>Output file scores.gistic from the run of GISTIC2.0</p></dd>


<dt id="arg-scores-df">scores_df<a class="anchor" aria-label="anchor" href="#arg-scores-df"></a></dt>
<dd><p>Optional. Instead of specifying scores_path pass a pre-loaded scores file as a data frame using scores_df</p></dd>


<dt id="arg-labels-bed">labels_bed<a class="anchor" aria-label="anchor" href="#arg-labels-bed"></a></dt>
<dd><p>Optional. A bed_data object specifying the regions to apply labels.</p></dd>


<dt id="arg-genome-build">genome_build<a class="anchor" aria-label="anchor" href="#arg-genome-build"></a></dt>
<dd><p>Defines the chr prefix and the coordinates of the default genes to label if `genes_to_label` is not provided.
Automatically set if labels_bed is provided</p></dd>


<dt id="arg-cutoff">cutoff<a class="anchor" aria-label="anchor" href="#arg-cutoff"></a></dt>
<dd><p>Optional. Used to determine which regions to color as aberrant. Must be float in the range between 0 and 1. The higher the number, the less regions will be considered as aberrant. The default is 0.5.</p></dd>


<dt id="arg-adjust-amps">adjust_amps<a class="anchor" aria-label="anchor" href="#arg-adjust-amps"></a></dt>
<dd><p>Optional. The value of G-score for highest amplification peak will be multiplied by this value to determine how far up the gene label will be displayed. Default 0.5.</p></dd>


<dt id="arg-adjust-dels">adjust_dels<a class="anchor" aria-label="anchor" href="#arg-adjust-dels"></a></dt>
<dd><p>Optional. The value of G-score for highest deletion peak will be multiplied by this value to determine how far down the gene label will be displayed. Default 2.75.</p></dd>


<dt id="arg-label-size">label_size<a class="anchor" aria-label="anchor" href="#arg-label-size"></a></dt>
<dd><p>Optional. The font size for the gene label to be displayed. Default 3.</p></dd>


<dt id="arg-force-pull">force_pull<a class="anchor" aria-label="anchor" href="#arg-force-pull"></a></dt>
<dd><p>Optional. How strong the gene name label will be pulled towards a data point. Default 0 (no pulling).</p></dd>


<dt id="arg-segment-curvature">segment.curvature<a class="anchor" aria-label="anchor" href="#arg-segment-curvature"></a></dt>
<dd><p>Optional. Indicates whether arrow to the data point should be curved. Accepts numeric value, where negative is for left-hand and positive for right-hand curves, and 0 for straight lines. Default 0.25.</p></dd>


<dt id="arg-segment-ncp">segment.ncp<a class="anchor" aria-label="anchor" href="#arg-segment-ncp"></a></dt>
<dd><p>Optional. Indicates number of control points to make a smoother curve. Higher value allows for more flexibility for the curve. Default 4.</p></dd>


<dt id="arg-segment-angle">segment.angle<a class="anchor" aria-label="anchor" href="#arg-segment-angle"></a></dt>
<dd><p>Optional. Numeric value in the range 0-180, where less than 90 skews control points of the arrow from label to data point toward the start point. Default 25.</p></dd>


<dt id="arg-hide-neutral">hide_neutral<a class="anchor" aria-label="anchor" href="#arg-hide-neutral"></a></dt>
<dd><p>Optional. Set to TRUE to hide all neutral (insignificant) regions instead of plotting them in grey</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>plot</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function uses GISTIC2.0 scores to create a chromosome plot, based on a similar plotting function from `maftools`.
The only required parameter for this function is `scores`, which is the path to a file with GISTIC2.0 scores.
Other parameters are all optional. For a detailed explanation of how to use these, refer to the parameter descriptions.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Bundled output from a GISTIC run using grch37 results</span></span></span>
<span class="r-in"><span><span class="va">gistic_scores</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"scores.gistic"</span>,</span></span>
<span class="r-in"><span>  package<span class="op">=</span><span class="st">"GAMBLR.viz"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># Automatic labeling of gene sets for a given pathology</span></span></span>
<span class="r-in"><span><span class="fu">prettyChromoplot</span><span class="op">(</span>scores_path <span class="op">=</span> <span class="va">gistic_scores</span>,</span></span>
<span class="r-in"><span>                 default_gene_set <span class="op">=</span> <span class="st">"FL"</span>,</span></span>
<span class="r-in"><span>                 genome_build <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>ggrepel: 17 unlabeled data points (too many overlaps). Consider increasing max.overlaps</span>
<span class="r-plt img"><img src="prettyChromoplot-1.png" alt="" width="864" height="778"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Specifying your own gene list for labeling</span></span></span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="st">"MYC"</span>,<span class="st">"FCGR2B"</span>,<span class="st">"TNFRSF14"</span>,<span class="st">"FAS"</span>,<span class="st">"PTEN"</span>,<span class="st">"B2M"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"RB1"</span>,<span class="st">"TCL1A"</span>,<span class="st">"CD70"</span>,<span class="st">"TOX"</span>,<span class="st">"PRAME"</span>,<span class="st">"CD38"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"BCL2"</span>,<span class="st">"KLHL14"</span>,<span class="st">"TCF4"</span>,<span class="st">"REL"</span>,<span class="st">"BCL6"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"SMARCA4"</span>,<span class="st">"CDKN2A"</span>,<span class="st">"RHOA"</span>,<span class="st">"HIST1H1C"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"TNFAIP3"</span>,<span class="st">"TP53"</span>,<span class="st">"CDK14"</span>,<span class="st">"RELN"</span>,<span class="st">"ETS1"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"MDM1"</span>,<span class="st">"MIR17HG"</span>,<span class="st">"CD58"</span>,<span class="st">"HNRNPD"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gene_bed</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GAMBLR.data/man/grch37_gene_coordinates.html" class="external-link">grch37_gene_coordinates</a></span>,<span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span><span class="co">#remove ensembl ID column</span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">hugo_symbol</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">genes</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span><span class="co">#keep genes of interest</span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>length <span class="op">=</span> <span class="va">end</span> <span class="op">-</span> <span class="va">start</span>,mid <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="va">length</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>start <span class="op">=</span> <span class="va">mid</span>,end<span class="op">=</span><span class="va">start</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span><span class="co">#convert to bed_data format</span></span></span>
<span class="r-in"><span> <span class="fu">GAMBLR.utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.utils/man/create_bed_data.html" class="external-link">create_bed_data</a></span><span class="op">(</span>genome_build <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">prettyChromoplot</span><span class="op">(</span>scores_path <span class="op">=</span> <span class="va">gistic_scores</span>,</span></span>
<span class="r-in"><span>                 labels_bed <span class="op">=</span> <span class="va">gene_bed</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="prettyChromoplot-2.png" alt="" width="864" height="778"></span>
<span class="r-in"><span><span class="co">#NOTE: genome build is inferred from gene_bed</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span> <span class="co"># GISTIC run using hg38 data</span></span></span>
<span class="r-in"><span><span class="fu">prettyChromoplot</span><span class="op">(</span>scores_path<span class="op">=</span><span class="va">gistic_scores</span>,</span></span>
<span class="r-in"><span>                   cutoff <span class="op">=</span> <span class="fl">0.9</span>,</span></span>
<span class="r-in"><span>                   label_size<span class="op">=</span><span class="fl">2</span>,</span></span>
<span class="r-in"><span>                   adjust_amps <span class="op">=</span> <span class="fl">0.5</span>,</span></span>
<span class="r-in"><span>                   adjust_dels <span class="op">=</span> <span class="fl">0.8</span>,</span></span>
<span class="r-in"><span>                   genome_build<span class="op">=</span><span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>                   hide_neutral <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

