<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Chromosome Plot — prettyChromoplot • GAMBLR.viz</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Chromosome Plot — prettyChromoplot"><meta property="og:description" content="Use GISTIC2.0 scores output to reproduce maftools::chromoplot with more flexibility."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">GAMBLR.viz</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Chromosome Plot</h1>
    
    <div class="hidden name"><code>prettyChromoplot.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Use GISTIC2.0 scores output to reproduce maftools::chromoplot with more flexibility.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">prettyChromoplot</span><span class="op">(</span></span>
<span>  <span class="va">scores_path</span>,</span>
<span>  <span class="va">scores_df</span>,</span>
<span>  <span class="va">labels_bed</span>,</span>
<span>  default_gene_set <span class="op">=</span> <span class="st">"oncogenes"</span>,</span>
<span>  <span class="va">genome_build</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  adjust_amps <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  adjust_dels <span class="op">=</span> <span class="fl">2.75</span>,</span>
<span>  label_size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  force_pull <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  segment.curvature <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  segment.ncp <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  segment.angle <span class="op">=</span> <span class="fl">25</span>,</span>
<span>  hide_neutral <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>scores_path</dt>
<dd><p>Output file scores.gistic from the run of GISTIC2.0</p></dd>


<dt>scores_df</dt>
<dd><p>Optional. Instead of specifying scores_path pass a pre-loaded scores file as a data frame using scores_df</p></dd>


<dt>labels_bed</dt>
<dd><p>Optional. A bed_data object specifying the regions to apply labels.</p></dd>


<dt>genome_build</dt>
<dd><p>Defines the chr prefix and the coordinates of the default genes to label if `genes_to_label` is not provided.
Automatically set if labels_bed is provided</p></dd>


<dt>cutoff</dt>
<dd><p>Optional. Used to determine which regions to color as aberrant. Must be float in the range between 0 and 1. The higher the number, the less regions will be considered as aberrant. The default is 0.5.</p></dd>


<dt>adjust_amps</dt>
<dd><p>Optional. The value of G-score for highest amplification peak will be multiplied by this value to determine how far up the gene label will be displayed. Default 0.5.</p></dd>


<dt>adjust_dels</dt>
<dd><p>Optional. The value of G-score for highest deletion peak will be multiplied by this value to determine how far down the gene label will be displayed. Default 2.75.</p></dd>


<dt>label_size</dt>
<dd><p>Optional. The font size for the gene label to be displayed. Default 3.</p></dd>


<dt>force_pull</dt>
<dd><p>Optional. How strong the gene name label will be pulled towards a data point. Default 0 (no pulling).</p></dd>


<dt>segment.curvature</dt>
<dd><p>Optional. Indicates whether arrow to the data point should be curved. Accepts numeric value, where negative is for left-hand and positive for right-hand curves, and 0 for straight lines. Default 0.25.</p></dd>


<dt>segment.ncp</dt>
<dd><p>Optional. Indicates number of control points to make a smoother curve. Higher value allows for more flexibility for the curve. Default 4.</p></dd>


<dt>segment.angle</dt>
<dd><p>Optional. Numeric value in the range 0-180, where less than 90 skews control points of the arrow from label to data point toward the start point. Default 25.</p></dd>


<dt>hide_neutral</dt>
<dd><p>Optional. Set to TRUE to hide all neutral (insignificant) regions instead of plotting them in grey</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>plot</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>This function uses GISTIC2.0 scores to create a chromosome plot, based on a similar plotting function from `maftools`.
The only required parameter for this function is `scores`, which is the path to a file with GISTIC2.0 scores.
Other parameters are all optional. For a detailed explanation of how to use these, refer to the parameter descriptions.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Bundled output from a GISTIC run using grch37 results</span></span></span>
<span class="r-in"><span><span class="va">gistic_scores</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"scores.gistic"</span>,</span></span>
<span class="r-in"><span>  package<span class="op">=</span><span class="st">"GAMBLR.viz"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># Automatic labeling of gene sets for a given pathology</span></span></span>
<span class="r-in"><span><span class="fu">prettyChromoplot</span><span class="op">(</span>scores_path <span class="op">=</span> <span class="va">gistic_scores</span>,</span></span>
<span class="r-in"><span>                 default_gene_set <span class="op">=</span> <span class="st">"FL"</span>,</span></span>
<span class="r-in"><span>                 genome_build <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>ggrepel: 17 unlabeled data points (too many overlaps). Consider increasing max.overlaps</span>
<span class="r-plt img"><img src="prettyChromoplot-1.png" alt="" width="700" height="433"></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Specifying your own gene list for labeling</span></span></span>
<span class="r-in"><span><span class="va">genes</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="st">"MYC"</span>,<span class="st">"FCGR2B"</span>,<span class="st">"TNFRSF14"</span>,<span class="st">"FAS"</span>,<span class="st">"PTEN"</span>,<span class="st">"B2M"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"RB1"</span>,<span class="st">"TCL1A"</span>,<span class="st">"CD70"</span>,<span class="st">"TOX"</span>,<span class="st">"PRAME"</span>,<span class="st">"CD38"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"BCL2"</span>,<span class="st">"KLHL14"</span>,<span class="st">"TCF4"</span>,<span class="st">"REL"</span>,<span class="st">"BCL6"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"SMARCA4"</span>,<span class="st">"CDKN2A"</span>,<span class="st">"RHOA"</span>,<span class="st">"HIST1H1C"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"TNFAIP3"</span>,<span class="st">"TP53"</span>,<span class="st">"CDK14"</span>,<span class="st">"RELN"</span>,<span class="st">"ETS1"</span>,</span></span>
<span class="r-in"><span>  <span class="st">"MDM1"</span>,<span class="st">"MIR17HG"</span>,<span class="st">"CD58"</span>,<span class="st">"HNRNPD"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">gene_bed</span> <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">grch37_gene_coordinates</span>,<span class="op">-</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span><span class="co">#remove ensembl ID column</span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">hugo_symbol</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">genes</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span> </span></span>
<span class="r-in"><span><span class="co">#keep genes of interest</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>length <span class="op">=</span> <span class="va">end</span> <span class="op">-</span> <span class="va">start</span>,mid <span class="op">=</span> <span class="va">start</span> <span class="op">+</span> <span class="va">length</span><span class="op">/</span><span class="fl">2</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>start <span class="op">=</span> <span class="va">mid</span>,end<span class="op">=</span><span class="va">start</span><span class="op">+</span><span class="fl">1</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span><span class="co">#convert to bed_data format</span></span></span>
<span class="r-in"><span>  <span class="fu">create_bed_data</span><span class="op">(</span>genome_build <span class="op">=</span> <span class="st">"grch37"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">prettyChromoplot</span><span class="op">(</span>scores_path <span class="op">=</span> <span class="va">gistic_scores</span>,</span></span>
<span class="r-in"><span>                 labels_bed <span class="op">=</span> <span class="va">gene_bed</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="prettyChromoplot-2.png" alt="" width="700" height="433"></span>
<span class="r-in"><span><span class="co">#NOTE: genome build is inferred from gene_bed</span></span></span>
<span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span> <span class="co"># GISTIC run using hg38 data</span></span></span>
<span class="r-in"><span><span class="fu">prettyChromoplot</span><span class="op">(</span>scores_path<span class="op">=</span><span class="va">gistic_scores</span>,</span></span>
<span class="r-in"><span>                   cutoff <span class="op">=</span> <span class="fl">0.9</span>,</span></span>
<span class="r-in"><span>                   label_size<span class="op">=</span><span class="fl">2</span>,</span></span>
<span class="r-in"><span>                   adjust_amps <span class="op">=</span> <span class="fl">0.5</span>,</span></span>
<span class="r-in"><span>                   adjust_dels <span class="op">=</span> <span class="fl">0.8</span>,</span></span>
<span class="r-in"><span>                   genome_build<span class="op">=</span><span class="st">"hg38"</span>,</span></span>
<span class="r-in"><span>                   hide_neutral <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

