<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Pretty mutation density heatmap — prettyMutationDensity • GAMBLR.viz</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Pretty mutation density heatmap — prettyMutationDensity"><meta name="description" content="Obtain a heatmap of mutation counts across sliding windows for multiple regions."><meta property="og:description" content="Obtain a heatmap of mutation counts across sliding windows for multiple regions."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GAMBLR.viz</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.3</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Introduction</h6></li>
    <li><a class="dropdown-item" href="../articles/why.html">GAMBLR.viz</a></li>
    <li><a class="dropdown-item" href="../articles/install.html"></a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Tutorials</h6></li>
    <li><a class="dropdown-item" href="../articles/recap_figures.html">Reproducing figures from GAMBL papers</a></li>
    <li><a class="dropdown-item" href="../articles/forestplot.html">Easy forest plots</a></li>
    <li><a class="dropdown-item" href="../articles/CopyNumberHeatmap.html">Visualizing copy number data as heatmaps</a></li>
    <li><a class="dropdown-item" href="../articles/data_q.html">Exploring data quality</a></li>
    <li><a class="dropdown-item" href="../articles/oncoplot.html">Upgrade your Oncoplots</a></li>
    <li><a class="dropdown-item" href="../articles/getting_started.html">Metadata and GAMBLR colours</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><h6 class="dropdown-header" data-toc-skip>Help</h6></li>
    <li><a class="dropdown-item" href="../articles/faq.html">Frequently Asked Qestions</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Pretty mutation density heatmap</h1>

      <div class="d-none name"><code>prettyMutationDensity.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Obtain a heatmap of mutation counts across sliding windows for multiple regions.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">prettyMutationDensity</span><span class="op">(</span></span>
<span>  regions_list <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  regions_bed <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  these_samples_metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  these_sample_ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  this_seq_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"genome"</span>, <span class="st">"capture"</span><span class="op">)</span>,</span>
<span>  <span class="va">maf_data</span>,</span>
<span>  <span class="va">mut_freq_matrix</span>,</span>
<span>  <span class="va">projection</span>,</span>
<span>  region_padding <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  drop_unmutated <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  metadataColumns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span><span class="op">)</span>,</span>
<span>  sortByMetadataColumns <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  expressionColumns <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  orientation <span class="op">=</span> <span class="st">"sample_rows"</span>,</span>
<span>  <span class="va">skip_regions</span>,</span>
<span>  <span class="va">only_regions</span>,</span>
<span>  customColours <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  naColour <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>  backgroundColour <span class="op">=</span> <span class="st">"transparent"</span>,</span>
<span>  slide_by <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  window_size <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  split_regions <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  min_count_per_bin <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  min_bin_recurrence <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  min_mut_tumour <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  region_fontsize <span class="op">=</span> <span class="fl">8</span>,</span>
<span>  clustering_distance_samples <span class="op">=</span> <span class="st">"euclidean"</span>,</span>
<span>  cluster_samples <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">cluster_rows_heatmap</span>,</span>
<span>  <span class="va">split_samples_kmeans</span>,</span>
<span>  show_row_names <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  show_column_names <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  cluster_regions <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  show_gene_colours <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  label_regions_by <span class="op">=</span> <span class="st">"name"</span>,</span>
<span>  merge_genes <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  label_regions_rotate <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  legend_row <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  legend_col <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  show_legend <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  legend_direction <span class="op">=</span> <span class="st">"horizontal"</span>,</span>
<span>  legendFontSize <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  metadataBarHeight <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  metadataBarFontsize <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  metadataSide <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>  region_annotation_name_side <span class="op">=</span> <span class="st">"top"</span>,</span>
<span>  sample_annotation_name_side <span class="op">=</span> <span class="st">"left"</span>,</span>
<span>  legend_side <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>  returnEverything <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  from_indexed_flatfile <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  mode <span class="op">=</span> <span class="st">"slms-3"</span>,</span>
<span>  <span class="va">width</span>,</span>
<span>  <span class="va">height</span>,</span>
<span>  hide_annotation_name <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  use_raster <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-regions-list">regions_list<a class="anchor" aria-label="anchor" href="#arg-regions-list"></a></dt>
<dd><p>Named vector of regions in the format c(name1 = "chr:start-end", name2 = "chr:start-end"). Only one (or none) between `regions_list` and `regions_bed` arguments should be provided. If neither regions_list nor regions_bed is specified, the function will use GAMBLR aSHM region information.</p></dd>


<dt id="arg-regions-bed">regions_bed<a class="anchor" aria-label="anchor" href="#arg-regions-bed"></a></dt>
<dd><p>Data frame of regions with four columns (chrom, start, end, name). Only one (or nome) between `regions_list` and `regions_bed` arguments should be provided.</p></dd>


<dt id="arg-these-samples-metadata">these_samples_metadata<a class="anchor" aria-label="anchor" href="#arg-these-samples-metadata"></a></dt>
<dd><p>Metadata with at least sample_id column. If not providing a maf data frame, seq_type is also required.</p></dd>


<dt id="arg-these-sample-ids">these_sample_ids<a class="anchor" aria-label="anchor" href="#arg-these-sample-ids"></a></dt>
<dd><p>Vector of sample IDs. Metadata will be subset to sample IDs present in this vector.</p></dd>


<dt id="arg-this-seq-type">this_seq_type<a class="anchor" aria-label="anchor" href="#arg-this-seq-type"></a></dt>
<dd><p>Optional vector of seq_types to include in heatmap. Default c("genome", "capture"). Uses default seq_type priority for samples with &gt;1 seq_type.</p></dd>


<dt id="arg-maf-data">maf_data<a class="anchor" aria-label="anchor" href="#arg-maf-data"></a></dt>
<dd><p>Optional maf data frame. Will be subset to rows where Tumor_Sample_Barcode matches provided sample IDs or metadata table. If not provided, maf data will be obtained with get_ssm_by_regions().</p></dd>


<dt id="arg-mut-freq-matrix">mut_freq_matrix<a class="anchor" aria-label="anchor" href="#arg-mut-freq-matrix"></a></dt>
<dd><p>Optional matrix of binned mutation frequencies generated outside of this function, usually by [GAMBLR::calc_mutation_frequency_bin_regions].</p></dd>


<dt id="arg-projection">projection<a class="anchor" aria-label="anchor" href="#arg-projection"></a></dt>
<dd><p>Genome build the function will operate in. Ensure this matches your provided regions and maf data for correct chr prefix handling. Default grch37.</p></dd>


<dt id="arg-region-padding">region_padding<a class="anchor" aria-label="anchor" href="#arg-region-padding"></a></dt>
<dd><p>Amount to pad the start and end coordinates by. Default 1000</p></dd>


<dt id="arg-drop-unmutated">drop_unmutated<a class="anchor" aria-label="anchor" href="#arg-drop-unmutated"></a></dt>
<dd><p>Whether to drop bins with 0 mutations. If returning a matrix format, this will only drop bins with no mutations in any samples.</p></dd>


<dt id="arg-metadatacolumns">metadataColumns<a class="anchor" aria-label="anchor" href="#arg-metadatacolumns"></a></dt>
<dd><p>Mandatory character vector of metadata columns to use in heatmap annotation. Default c("pathology").</p></dd>


<dt id="arg-sortbymetadatacolumns">sortByMetadataColumns<a class="anchor" aria-label="anchor" href="#arg-sortbymetadatacolumns"></a></dt>
<dd><p>Optional character vector of metadata columns to order annotations by. Will be ordered by factor levels and sorted in the order specified. Default NULL.</p></dd>


<dt id="arg-expressioncolumns">expressionColumns<a class="anchor" aria-label="anchor" href="#arg-expressioncolumns"></a></dt>
<dd><p>Optional character vector of numeric metadata columns, usually gene expression, for heatmap annotation.</p></dd>


<dt id="arg-orientation">orientation<a class="anchor" aria-label="anchor" href="#arg-orientation"></a></dt>
<dd><p>Specify whether heatmap should have samples in rows ("sample_rows") or in columns ("sample_cols"). Default sample_rows.</p></dd>


<dt id="arg-skip-regions">skip_regions<a class="anchor" aria-label="anchor" href="#arg-skip-regions"></a></dt>
<dd><p>Optional character vector of genes to exclude from the default aSHM regions.</p></dd>


<dt id="arg-only-regions">only_regions<a class="anchor" aria-label="anchor" href="#arg-only-regions"></a></dt>
<dd><p>Optional character vector of genes to include from the default aSHM regions.</p></dd>


<dt id="arg-customcolours">customColours<a class="anchor" aria-label="anchor" href="#arg-customcolours"></a></dt>
<dd><p>Optional list of character vectors specifying colours for heatmap annotation with metadataColumns, e.g. list(pathology = c(DLBCL = "green", BL = "purple")). If left blank, the function will attempt to match heatmap annotations with existing colours from [GAMBLR::get_gambl_colours], or will default to the Blood colour palette.</p></dd>


<dt id="arg-nacolour">naColour<a class="anchor" aria-label="anchor" href="#arg-nacolour"></a></dt>
<dd><p>Colour to use for NA values in metadata/expression. Default "white".</p></dd>


<dt id="arg-backgroundcolour">backgroundColour<a class="anchor" aria-label="anchor" href="#arg-backgroundcolour"></a></dt>
<dd><p>Optionally specify the colour for heatmap bins with 0 mutations. Default grey90.</p></dd>


<dt id="arg-slide-by">slide_by<a class="anchor" aria-label="anchor" href="#arg-slide-by"></a></dt>
<dd><p>Slide size for sliding window. Default 100.</p></dd>


<dt id="arg-window-size">window_size<a class="anchor" aria-label="anchor" href="#arg-window-size"></a></dt>
<dd><p>Size of sliding window. Default 500.</p></dd>


<dt id="arg-min-count-per-bin">min_count_per_bin<a class="anchor" aria-label="anchor" href="#arg-min-count-per-bin"></a></dt>
<dd><p>Specify the minimum number of mutations per bin to be included in the heatmap. Only bins with all samples falling below this threshold will be dropped. Default 0.</p></dd>


<dt id="arg-min-bin-recurrence">min_bin_recurrence<a class="anchor" aria-label="anchor" href="#arg-min-bin-recurrence"></a></dt>
<dd><p>Specify how many samples a bin must be mutated in to be displayed. Default 5.</p></dd>


<dt id="arg-min-mut-tumour">min_mut_tumour<a class="anchor" aria-label="anchor" href="#arg-min-mut-tumour"></a></dt>
<dd><p>Specify how many bins a tumour must be mutated in to be displayed. Default 0.</p></dd>


<dt id="arg-region-fontsize">region_fontsize<a class="anchor" aria-label="anchor" href="#arg-region-fontsize"></a></dt>
<dd><p>Fontsize of region labels on the heatmap. Default 8.</p></dd>


<dt id="arg-show-gene-colours">show_gene_colours<a class="anchor" aria-label="anchor" href="#arg-show-gene-colours"></a></dt>
<dd><p>Boolean. Whether to add heatmap annotation colours for each region. Default FALSE.</p></dd>


<dt id="arg-label-regions-by">label_regions_by<a class="anchor" aria-label="anchor" href="#arg-label-regions-by"></a></dt>
<dd><p>Specify which feature of the regions to label the heatmap with. Heatmap will be split according to this value, and ordered by factor levels if the specified column is a factor. Default name.</p></dd>


<dt id="arg-merge-genes">merge_genes<a class="anchor" aria-label="anchor" href="#arg-merge-genes"></a></dt>
<dd><p>Set to TRUE to drop everything after "-" in the label to collpse regions from the same gene/locus. Default FALSE.</p></dd>


<dt id="arg-label-regions-rotate">label_regions_rotate<a class="anchor" aria-label="anchor" href="#arg-label-regions-rotate"></a></dt>
<dd><p>Specify degree by which the label in the previous parameter will be rotated. Default 0 (no rotation). The accepted values are 0, 90, 270.</p></dd>


<dt id="arg-legend-row">legend_row<a class="anchor" aria-label="anchor" href="#arg-legend-row"></a></dt>
<dd><p>Control aesthetics of the heatmap legend. Default 3.</p></dd>


<dt id="arg-legend-col">legend_col<a class="anchor" aria-label="anchor" href="#arg-legend-col"></a></dt>
<dd><p>Control aesthetics of the heatmap legend. Default 3.</p></dd>


<dt id="arg-show-legend">show_legend<a class="anchor" aria-label="anchor" href="#arg-show-legend"></a></dt>
<dd><p>Boolean. Default TRUE.</p></dd>


<dt id="arg-legend-direction">legend_direction<a class="anchor" aria-label="anchor" href="#arg-legend-direction"></a></dt>
<dd><p>Control aesthetics of the heatmap legend. Default "horizontal".</p></dd>


<dt id="arg-legendfontsize">legendFontSize<a class="anchor" aria-label="anchor" href="#arg-legendfontsize"></a></dt>
<dd><p>Control aesthetics of the heatmap legend. Default 10.</p></dd>


<dt id="arg-metadatabarheight">metadataBarHeight<a class="anchor" aria-label="anchor" href="#arg-metadatabarheight"></a></dt>
<dd><p>Optional argument to adjust the height of bar with annotations. The default is 1.5.</p></dd>


<dt id="arg-metadatabarfontsize">metadataBarFontsize<a class="anchor" aria-label="anchor" href="#arg-metadatabarfontsize"></a></dt>
<dd><p>Optional argument to control for the font size of metadata annotations. The default is 5.</p></dd>


<dt id="arg-metadataside">metadataSide<a class="anchor" aria-label="anchor" href="#arg-metadataside"></a></dt>
<dd><p>Default location for metadata is the bottom. Set to "top" if you want to move it</p></dd>


<dt id="arg-legend-side">legend_side<a class="anchor" aria-label="anchor" href="#arg-legend-side"></a></dt>
<dd><p>Control aesthetics of the heatmap legend. Default "bottom".</p></dd>


<dt id="arg-returneverything">returnEverything<a class="anchor" aria-label="anchor" href="#arg-returneverything"></a></dt>
<dd><p>Boolean. FALSE will plot the heatmap automatically. TRUE will return a heatmap object to allow further tweaking with the draw() function. Default FALSE.</p></dd>


<dt id="arg-from-indexed-flatfile">from_indexed_flatfile<a class="anchor" aria-label="anchor" href="#arg-from-indexed-flatfile"></a></dt>
<dd><p>Set to TRUE to avoid using the database and instead rely on flat files (only works for streamlined data, not full MAF details).</p></dd>


<dt id="arg-mode">mode<a class="anchor" aria-label="anchor" href="#arg-mode"></a></dt>
<dd><p>Only works with indexed flat files. Accepts 2 options of "slms-3" and "strelka2" to indicate which variant caller to use. Default is "slms-3".</p></dd>


<dt id="arg-use-raster">use_raster<a class="anchor" aria-label="anchor" href="#arg-use-raster"></a></dt>
<dd><p>Control whether ComplexHeatmap uses rastering. default: TRUE</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A table of mutation counts for sliding windows across one or more regions. May be long or wide.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>This function takes a metadata table with `these_samples_metadata` parameter and internally calls [GAMBLR::calc_mutation_frequency_bin_region] (that internally calls [GAMBLR::get_ssm_by_regions]).
to retrieve mutation counts for sliding windows across one or more regions and generate a heatmap. May optionally provide any combination of a maf data frame, existing metadata, or a regions data frame or named vector.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GAMBLR.open</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># get meta data</span></span></span>
<span class="r-in"><span><span class="va">my_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GAMBLR.open/man/get_gambl_metadata.html" class="external-link">get_gambl_metadata</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pathology</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FL"</span>,<span class="st">"DLBCL"</span><span class="op">)</span>, <span class="va">seq_type</span> <span class="op">!=</span> <span class="st">"mrna"</span><span class="op">)</span> <span class="op"><a href="pipe.html">%&gt;%</a></span></span></span>
<span class="r-in"><span>  <span class="fu">check_and_clean_metadata</span><span class="op">(</span>duplicate_action <span class="op">=</span> <span class="st">"keep_first"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Using the bundled metadata in GAMBLR.data...</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Duplicate rows (keeping first occurrence) for 'sample_id' and 'seq_type' have been dropped.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># get ashm regions of a set of genes.</span></span></span>
<span class="r-in"><span><span class="va">my_regions</span> <span class="op">=</span> <span class="fu">create_bed_data</span><span class="op">(</span><span class="fu">GAMBLR.data</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/GAMBLR.data/man/grch37_ashm_regions.html" class="external-link">grch37_ashm_regions</a></span>,</span></span>
<span class="r-in"><span>  fix_names <span class="op">=</span> <span class="st">"concat"</span>,</span></span>
<span class="r-in"><span>  concat_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene"</span>,<span class="st">"region"</span><span class="op">)</span>,sep <span class="op">=</span> <span class="st">"-"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># create heatmap of mutation counts for the specified regions</span></span></span>
<span class="r-in"><span><span class="va">meta_columns</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pathology"</span>,</span></span>
<span class="r-in"><span>                  <span class="st">"lymphgen"</span>,</span></span>
<span class="r-in"><span>                  <span class="st">"COO_consensus"</span>, </span></span>
<span class="r-in"><span>                  <span class="st">"DHITsig_consensus"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="op">{</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu">prettyMutationDensity</span><span class="op">(</span></span></span>
<span class="r-in"><span>   regions_bed <span class="op">=</span> <span class="va">my_regions</span>,</span></span>
<span class="r-in"><span>   these_samples_metadata <span class="op">=</span> <span class="va">my_meta</span>,</span></span>
<span class="r-in"><span>   metadataColumns <span class="op">=</span> <span class="va">meta_columns</span>,</span></span>
<span class="r-in"><span>   orientation<span class="op">=</span><span class="st">"sample_columns"</span>,</span></span>
<span class="r-in"><span>   sortByMetadataColumns <span class="op">=</span> <span class="va">meta_columns</span>,</span></span>
<span class="r-in"><span>   projection <span class="op">=</span> <span class="st">"grch37"</span>,</span></span>
<span class="r-in"><span>   backgroundColour <span class="op">=</span> <span class="st">"transparent"</span>,</span></span>
<span class="r-in"><span>   show_legend <span class="op">=</span> <span class="cn">FALSE</span>,</span></span>
<span class="r-in"><span>   region_fontsize <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-plt img"><img src="prettyMutationDensity-1.png" alt="" width="864" height="778"></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Ryan Morin, Kostiantyn Dreval, Laura Hilton, Adam Mattsson, Vladimir Souza.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

