#' @title Pretty MAF summary
#'
#' @description Generates a visually appealling alternative to
#' MAFtools `plotmafSummary``.
#'
#' @details This function reproduces the multi-panel MAF summary plot
#' generated by MAFtools. The colour, plot aesthetics etc utilize GAMBLR colours
#' and are consistent with other GAMBLR.viz plots such as prettyOncoplot.
#'
#' @param maf_data A data frame containing the mutation data.
#' @param base_size Base size of fonts to use in Morons theme
#' @param top_n Number of genes to show in top_n panel (default 10)
#' @param point_size Size of points to use in ggbeeswarm plot (default 0.5)
#' @param returnEverything Set to TRUE to get the underlying numbers
#' and all panels as a list
#'
#' @return A plot
#'
#' @import dplyr ggplot2 cowplot ggbeeswarm GAMBLR.utils
#' @export
#'
#' @examples
#' suppressMessages(library(GAMBLR.open))
#' FL_meta <- GAMBLR.open::get_gambl_metadata() %>%
#'     dplyr::filter(pathology == "FL") %>%
#'     check_and_clean_metadata(.,duplicate_action="keep_first")
#' FL_coding <- GAMBLR.open::get_coding_ssm(these_samples_metadata = FL_meta)
#'
#' pretty_MAF_summary(FL_coding)
#'
#' pretty_MAF_summary(FL_coding, top_n = 22, base_size = 6)
#'
pretty_MAF_summary <- function(maf_data,
                              these_samples_metadata,
                              base_size = 4,
                              top_n = 10,
                              point_size = 0.5,
                              one_per_sample = FALSE,
                              gene_sizes,
                              returnEverything = FALSE) {
  p1 <- make_panel1(maf_data,
    base_size = base_size, title = "Variant Classification"
  )
  p2 <- make_panel2(maf_data,
    base_size = base_size, title = "Variant Type"
  )
  p3 <- make_panel3(maf_data,
    base_size = base_size, title = "SNV Class"
  )
  p4 <- make_panel4(maf_data,
    base_size = base_size, title = "Variants per sample"
  )
  p5 <- make_panel5(maf_data,
    base_size = base_size,
    point_size = point_size,
    title = "Variant Classification Summary"
  )
  if(one_per_sample){
    title6 <- paste("Top", top_n, "genes (per sample)")
  }else{
    title6 <- paste("Top", top_n, "genes (cumulative)")
  }
  p6_list <- make_panel6(maf_data,
    top = top_n,
    base_size = base_size,
    title = title6,
    one_per_sample = one_per_sample 
  )
  p6 = p6_list$plot
  if(!missing(gene_sizes)){
    #subset to these_samples_metadata
    samples = these_samples_metadata$sample_id
    p6_background = make_panel6(dplyr::filter(maf_data,
                                              !Tumor_Sample_Barcode %in% samples),
                                top = top_n,
                                base_size = base_size,
                                title = title6,
                                one_per_sample = one_per_sample 
    )
    p6_foreground = make_panel6(dplyr::filter(maf_data,
                                              Tumor_Sample_Barcode %in% samples),
                                top = top_n,
                                base_size = base_size,
                                title = title6,
                                one_per_sample = one_per_sample 
    )
    print("Panel 7")
    p7_out <- make_panel7(strip_genomic_classes(p6_foreground$tabulated),
                         strip_genomic_classes(p6_background$tabulated),
                         gene_sizes)

    
  }
  
  all_p <- cowplot::plot_grid(p1, p2, p3, p4, p5, p6, nrow = 2, ncol = 3)
  if(returnEverything){
    to_ret = list(gene_freq = GAMBLR.utils::strip_genomic_classes(p6_list$tabulated),
                combined=all_p,
                panel1=p1,
                panel2=p2,
                panel3=p3,
                panel4=p4,
                panel5=p5,
                panel6=p6)
    if(!missing(gene_sizes)){
        #for(p7 in names(p7_out)){
        #    to_ret[[p7]] = p7_out[[p7]]
        #}
        to_ret[["tabulated"]] = p7_out$tabulated
    }
    return(to_ret)
  }
  all_p
}

#' @keywords internal
make_panel1 <- function(maf_data, base_size = 5, title = "") {
    vc_counted <- maf_data %>%
        group_by(Variant_Classification) %>%
        count() %>%
        arrange(n)
    vc_counted$Variant_Classification <- factor(vc_counted$Variant_Classification,
        levels = unique(vc_counted$Variant_Classification)
    )
    mut_cols <- get_gambl_colours("mutation")
    p1 <- ggplot(vc_counted, aes(x = n, y = Variant_Classification, fill = Variant_Classification)) +
        geom_col() +
        scale_fill_manual(values = mut_cols) +
        theme_Morons(base_size = base_size, my_legend_position = "none") +
        theme(
            axis.title.x = element_blank(),
            axis.title.y = element_blank(),
            axis.text.x = element_blank(),
            axis.ticks.x = element_blank(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank()
        ) +
        ggtitle(title)
    p1
}



#' @keywords internal
make_panel2 <- function(maf_data, base_size = 5, title = "") {
    type_counted <- maf_data %>%
        group_by(Variant_Type) %>%
        count() %>%
        arrange(n)
    type_counted$Variant_Type <- factor(type_counted$Variant_Type,
        levels = unique(type_counted$Variant_Type)
    )

    mut_cols <- c(SNP = "purple1", INS = "yellow3", DEL = "lightblue", DNP = "orange", "TNP" = "lightgreen")
    p2 <- ggplot(type_counted, aes(x = n, y = Variant_Type, fill = Variant_Type)) +
        geom_col() +
        scale_fill_manual(values = mut_cols) +
        theme_Morons(base_size = base_size, my_legend_position = "none") +
        theme(
            axis.title.x = element_blank(),
            axis.title.y = element_blank(),
            axis.text.x = element_blank(),
            axis.ticks.x = element_blank(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank()
        ) +
        ggtitle(title)
    p2
}



#' @keywords internal
make_panel3 <- function(maf_data, base_size = 5, title = "") {
    comp <- function(base) {
        chartr("ACTG", "TGAC", base)
    }
    maf_data <- mutate(maf_data,
        class = case_when(
            Reference_Allele %in% c("T", "C") ~
                paste0(
                    Reference_Allele,
                    ">",
                    Tumor_Seq_Allele2
                ),
            TRUE ~ paste0(
                comp(Reference_Allele),
                ">",
                comp(Tumor_Seq_Allele2)
            )
        )
    )

    class_counted <- maf_data %>%
        dplyr::filter(Variant_Type == "SNP") %>%
        group_by(class) %>%
        count()
    class_counted <- mutate(class_counted, class = factor(class, levels = c("C>A", "C>G", "C>T", "T>C", "T>A", "T>G")))
    mut_cols <- get_gambl_colours("rainfall")
    p3 <- ggplot(class_counted, aes(x = n, y = class, fill = class)) +
        geom_col() +
        scale_fill_manual(values = mut_cols) +
        theme_Morons(base_size = base_size, my_legend_position = "none") +
        theme(
            axis.title.x = element_blank(),
            axis.title.y = element_blank(),
            axis.text.x = element_blank(),
            axis.ticks.x = element_blank(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank()
        ) +
        ggtitle(title)
    p3
}


#' @keywords internal
make_panel4 <- function(maf_data, base_size = 5, title = "") {
    type_counted <- maf_data %>%
        group_by(Tumor_Sample_Barcode, Variant_Classification) %>%
        count() %>%
        arrange(desc(n))
    type_counted$Tumor_Sample_Barcode <- factor(type_counted$Tumor_Sample_Barcode,
        levels = unique(type_counted$Tumor_Sample_Barcode)
    )

    mut_cols <- get_gambl_colours("mutation")
    p4 <- ggplot(type_counted, aes(x = Tumor_Sample_Barcode, y = n, fill = Variant_Classification)) +
        geom_col() +
        scale_fill_manual(values = mut_cols) +
        theme_Morons(base_size = base_size, my_legend_position = "none") +
        theme(
            axis.title.x = element_blank(),
            axis.title.y = element_blank(),
            axis.text.x = element_blank(),
            axis.ticks.x = element_blank(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank()
        ) +
        ggtitle(title)


    p4
}
#' @keywords internal
make_panel5 <- function(maf_data, base_size = 5, point_size = 0.5, title = "") {
    type_counted <- maf_data %>%
        group_by(Tumor_Sample_Barcode, Variant_Classification) %>%
        count() %>%
        arrange(desc(n))
    mut_cols <- get_gambl_colours("mutation")
    vc_counted <- maf_data %>%
        group_by(Variant_Classification) %>%
        count() %>%
        arrange(n)
    vc_counted$Variant_Classification <- factor(vc_counted$Variant_Classification,
        levels = unique(vc_counted$Variant_Classification)
    )
    type_counted$Variant_Classification <- factor(type_counted$Variant_Classification,
        levels = rev(unique(vc_counted$Variant_Classification))
    )
    p5 <- ggplot(type_counted, aes(x = Variant_Classification, y = n, colour = Variant_Classification)) +
        geom_quasirandom(size = point_size) +
        scale_colour_manual(values = mut_cols) +
        theme_Morons(base_size = base_size, my_legend_position = "none") +
        theme(
            axis.title.y = element_blank(),
            axis.text.x = element_blank(),
            axis.title.x = element_blank(),
            axis.ticks.x = element_blank(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank()
        ) +
        ggtitle(title)
    p5
}
#' @keywords internal
make_panel6 <- function(maf_data, base_size = 6, top = 10, title = "", one_per_sample = FALSE) {
    if(one_per_sample){
        #only one mutation per gene is counted per patient
        
        maf_data = maf_data %>%
            slice_head(n=1, by=c(Tumor_Sample_Barcode, Hugo_Symbol)) 
    }
    type_counted <-  maf_data %>%
        group_by(Hugo_Symbol, Variant_Classification) %>%
        count() %>%
        arrange(n)
    mut_cols <- get_gambl_colours("mutation")
    top_n <- group_by(type_counted, Hugo_Symbol) %>%
        summarise(total = sum(n)) %>%
        arrange(desc(total)) %>%
        slice_head(n = top) %>%
        pull(Hugo_Symbol)
    
    some_type_counted <- dplyr::filter(type_counted, Hugo_Symbol %in% top_n)
    some_type_counted$Hugo_Symbol <- factor(some_type_counted$Hugo_Symbol,
        levels = rev(top_n))
    p6 <-
        ggplot(some_type_counted, aes(y = Hugo_Symbol, x = n, fill = Variant_Classification)) +
        geom_col() +
        scale_fill_manual(values = mut_cols) +
        theme_Morons(base_size = base_size, my_legend_position = "none") +
        theme(
            axis.text.x = element_blank(),
            axis.title.y = element_blank(),
            axis.title.x = element_blank(),
            axis.ticks.x = element_blank(),
            panel.grid.major = element_blank(),
            panel.grid.minor = element_blank()
        ) +
        ggtitle(title)

    return(list(plot=p6, tabulated=type_counted))
}

make_panel7 <- function(mutcounts_these,
                        mutcounts_all,
                        gene_sizes){
  # use all samples in the MAF that aren't in these_samples_metadata as the background
  # use all samples in these_samples_metadata as the foreground
  # scale mutation frequency to the gene CDS length
  # e.g. as mutations per 1000 codons, per_kilocodon
  # separately consider Silent and non-silent
  coding_counts = dplyr::filter(mutcounts_these,
                                Variant_Classification!="Silent") %>%
    summarise(total_coding=sum(n), .by=Hugo_Symbol)
  
  silent_counts = dplyr::filter(mutcounts_all,
                                Variant_Classification=="Silent") %>%
    dplyr::rename("total_noncoding" = "n") %>% 
    dplyr::select(-Variant_Classification)
  all_counts = full_join(coding_counts,silent_counts,by="Hugo_Symbol") 
  all_counts = left_join(all_counts,gene_sizes) %>% 
    mutate(total_noncoding = ifelse(is.na(total_noncoding),0,total_noncoding)) %>%
    mutate(coding_per_kilocodon = total_coding * 1000 / CDS) %>%
    mutate(noncoding_per_kilocodon = total_noncoding * 1000 / CDS)
all_counts 
  return(list(plot=all_counts,tabulated=all_counts))
}
