utils::globalVariables(unique(c(
  "cn_col_seg", "cohort", "collate_results", "color", "colour", 
  "comparison", "conf.high", "conf.low", "coords", "cytobands_grch37", 
  "cytobands_hg38", "dev.off", "down", "end.points", "end1", "end2", 
  "end_col_maf", "end_col_seg", "endpos", "ensembl_gene_id", "estimate", 
  "fisher.test", "formalArgs", "fractMutated", "frequency", "fusion", 
  "gene", "gene_name", "gene_norm", "gene_total", "get_cn_segments", 
  "get_gambl_metadata", "get_manta_sv", "get_sample_cn_segments", 
  "get_ssm_by_region", "get_ssm_by_regions", "get_ssm_by_sample", 
  "get_ssm_by_samples", "grch37_ashm_regions", "group", "head", 
  "hg38_ashm_regions", "hot_spot", "hugo_symbol", "importance", 
  "incoming_cn", "is_mutated", "label", "left_distance", "lenght", 
  "letter", "local.maxima", "local.minima", "log.ratio", "lymphgen", 
  "major_cn", "manta_name", "method", "midpoint", "minor_cn", "mut", 
  "mut_status_rn", "mutate.genomic_data", "mutation", 
  "mutation_context", "mutation_count", "n_mutations", "name", 
  "newcol", "p.adjust", "p.value", "partner", "pathology", 
  "pathology_rank", "pdf", "percent_mutated", "protein_domains", 
  "q.value", "rainbow", "read.table", "ref", "region", "region_name", 
  "regions", "regions_maf", "regions_sub", "rel_start", 
  "right_distance", "sample_id", "seq_type", "setNames", "size", 
  "start.points", "start1", "start2", "start_col_maf", "start_col_seg", 
  "startpos", "str_detect", "str_remove", "t_alt_count", "t_depth", 
  "t_ref_count", "test", "text", "text.label", "text.position", "time", 
  "time.lo", "time.up", "total", "tumour_sample_id", "type", 
  "type_table", "up", "vaf", "value", "variable", 
  "variant_type_col_maf", "x", "ycoord", "ystart",
  ".",
  "AA", "CHROM_A", "CHROM_B", "CN", "CN2", "CN_segment", "Chromosome",
  "END_A", "END_B", "End", "End_Position", "Feature", "Freq", "G-score", 
  "Gene", "HGNC", "HGVSp_Short", "Hugo_Symbol", "ID", "IMD", "LOH_flag", 
  "MeanCorrectedCoverage", "Metric", "Mid", "MutatedSamples", "NA", 
  "N_features", "PGA", "ProportionCoverage10x", "ProportionCoverage30x", 
  "ProportionReadsDuplicated", "ProportionReadsMapped", "Reference_Allele", 
  "START_A", "START_B", "Size", "Start", "Start_Position", "Substitution", 
  "TotalDuplicatedreads", "TotalReads", "TotalUniquelyMapped", 
  "Tumor_Sample_Barcode", "Tumor_Seq_Allele2", "Type", "V1", "V2", "V3", 
  "V4", "VAF", "VAF_tumour", "Value", "Variant_Classification", 
  "Variant_Type", "WRCY", "aggregate", "alt", "alternative", 
  "annotate_ssm_motif_context", "arm", "assign_cn_to_ssm", "bin_name", 
  "calc_mutation_frequency_bin_regions", "cb.chromosome", "cb.end", 
  "cb.name", "cb.start", "centre", "chr", "chrom", "chrom1", "chrom2", 
  "chrom_col_seg", "chromosome", "chromosomeN", "chromosome_col_maf", 
  "chroms_u_nochr", "classification", "clonal_frequency", "cls", 
  "cn_0", "cn_0_sample1", "cn_0_sample2", "cn_0_sample3", 
  "cn_0_sample4", "cn_1", "cn_1_sample1", "cn_1_sample2", 
  "cn_1_sample3", "cn_1_sample4", "cn_2_sample1", "cn_2_sample2", 
  "cn_2_sample3", "cn_2_sample4", "cn_3", "cn_3_sample1", 
  "cn_3_sample2", "cn_3_sample3", "cn_3_sample4", "cn_4", 
  "cn_4_sample1", "cn_4_sample2", "cn_4_sample3", "cn_4_sample4", 
  "cn_5", "cn_5_sample1", "cn_5_sample2", "cn_5_sample3", 
  "cn_5_sample4", "cn_6", "cn_6_sample1", "cn_6_sample2", 
  "cn_6_sample3", "cn_6_sample4"
)))
